{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import seaborn as sns\n",
    "from skbio.diversity import alpha_diversity\n",
    "from skbio.diversity import beta_diversity\n",
    "from skbio import TreeNode\n",
    "from io import StringIO\n",
    "from skbio.stats.ordination import pcoa\n",
    "\n",
    "import cx_Oracle\n",
    "\n",
    "SERVER = 'dboracle.itk.ppke.hu'\n",
    "PORT = 1521\n",
    "SERVICE = 'rsc.itk.ppke.hu'\n",
    "USER = 'metagenome'\n",
    "PASSWORD = '.'\n",
    "\n",
    "class Connection():\n",
    "\n",
    "    def __init__(self):\n",
    "        dsn_tns = cx_Oracle.makedsn(SERVER, PORT, service_name=SERVICE)\n",
    "        self.connection = cx_Oracle.connect(USER, PASSWORD, dsn_tns)\n",
    "\n",
    "    def get_connection(self):\n",
    "        return self.connection\n",
    "\n",
    "    def select(self, sql):\n",
    "        return pd.read_sql(sql, con=self.connection)\n",
    "\n",
    "con = Connection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '//gfs/data/curated_metagenomes_kraken2/'\n",
    "\n",
    "files = {}\n",
    "# r=root, d=directories, f = files\n",
    "for r, d, f in os.walk(path):\n",
    "    for directory in d:\n",
    "        for r1, d1, f1 in os.walk(os.path.join(path,directory)):\n",
    "            for file in f1:\n",
    "                if '.report.out' in file:\n",
    "                    if directory in files:\n",
    "                        files[directory].append(os.path.join(r1, file))\n",
    "                    else:\n",
    "                        files[directory] = [(os.path.join(r1, file))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_data_for_sample(path_to_file):\n",
    "    dataset = path_to_file.split('/')[-2]\n",
    "    df = pd.read_csv(path_to_file, delimiter = '\\t', \n",
    "                     names=[\"Abundance\", \"Root_count\", \"Taxon_count\", \"Rank\", \"NCBI_ID\", \"Name\"])\n",
    "    sample_id = path_to_file.split('/')[-1].split('.')[0]\n",
    "    df['Sample_ID'] = sample_id\n",
    "    df['Dataset'] = dataset\n",
    "    not_found_files = set()\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def join_metadata(df, dataset_name):\n",
    "    SQL_METADATA = \"\"\"\n",
    "        select\n",
    "            DATASET_NAME,\n",
    "            STUDY_CONDITION,\n",
    "            BODY_SITE,\n",
    "            SUBJECTID,\n",
    "            NCBI_ACCESSION,\n",
    "            COUNTRY,\n",
    "            DISEASE,\n",
    "            GENDER,\n",
    "            AGE,\n",
    "            AGE_CATEGORY,\n",
    "            NUMBER_READS\n",
    "        from METADATA\n",
    "        where dataset_name = '{dataset_name}'\n",
    "        \"\"\".replace(\"{dataset_name}\", dataset_name)\n",
    "    df_metadata = con.select(SQL_METADATA)\n",
    "    \n",
    "    df[\"DATASET_NAME\"] = df[\"Dataset\"]\n",
    "    merged = df.merge(df_metadata, on='DATASET_NAME')\n",
    "    \n",
    "    col_to_idx = {}\n",
    "    for i in range(len(merged.columns)):\n",
    "        col_to_idx[merged.columns[i]] = i\n",
    "    \n",
    "    joined = merged.loc[[row[1][col_to_idx[\"Sample_ID\"]] in row[1][col_to_idx[\"NCBI_ACCESSION\"]] \n",
    "                        for row in merged.iterrows()]]\n",
    "    \n",
    "    output = joined[[\"Sample_ID\", \"Dataset\", \"Rank\", \"Name\", \"Abundance\", \"\", \"Taxon_count\", \"STUDY_CONDITION\", \n",
    "                     \"BODY_SITE\", \"SUBJECTID\", \"COUNTRY\", \"DISEASE\", \"AGE\", \"AGE_CATEGORY\", \"NUMBER_READS\"]]\n",
    "    output = output.reset_index()\n",
    "    output = output.drop(columns=\"index\")\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RampelliS_2015  kész.\n",
      "AsnicarF_2017  kész.\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-56-e54601c35305>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mdf_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mfile\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfilelist\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m             \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_data_for_sample\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m                 \u001b[0mdf_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_data_for_sample\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-54-82aabb654ad7>\u001b[0m in \u001b[0;36mget_data_for_sample\u001b[0;34m(path_to_file)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mdataset\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpath_to_file\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     df = pd.read_csv(path_to_file, delimiter = '\\t', \n\u001b[0;32m----> 4\u001b[0;31m                      names=[\"Abundance\", \"Root_count\", \"Taxon_count\", \"Rank\", \"NCBI_ID\", \"Name\"])\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0msample_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpath_to_file\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Sample_ID'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msample_id\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gfs/progs/miniconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, doublequote, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    676\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    677\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 678\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    679\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    680\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gfs/progs/miniconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    438\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    439\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 440\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    441\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    442\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gfs/progs/miniconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    785\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    786\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 787\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    788\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    789\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gfs/progs/miniconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1012\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'c'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1013\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'c'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1014\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1015\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1016\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'python'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gfs/progs/miniconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1706\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'usecols'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1707\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1708\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1709\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1710\u001b[0m         \u001b[0mpassed_names\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnames\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "dataset_dfs = {}\n",
    "result = joined\n",
    "for dataset_name, filelist in files.items():\n",
    "    if dataset_name = \"AsnicarF_2017\":\n",
    "        df_list = []\n",
    "        for file in filelist:\n",
    "            df = get_data_for_sample(file)\n",
    "            if df is not None:\n",
    "                df_list.append(get_data_for_sample(file))\n",
    "        df_dataset = pd.concat(df_list)\n",
    "        joined = join_metadata(df_dataset, dataset_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Percentage of fragments covered by the clade rooted at this taxon\n",
    "2. Number of fragments covered by the clade rooted at this taxon\n",
    "3. Number of fragments assigned directly to this taxon\n",
    "4. A rank code, indicating (U)nclassified, (R)oot, (D)omain, (K)ingdom,\n",
    "   (P)hylum, (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies.\n",
    "   Taxa that are not at any of these 10 ranks have a rank code that is\n",
    "   formed by using the rank code of the closest ancestor rank with\n",
    "   a number indicating the distance from that rank.  E.g., \"G2\" is a\n",
    "   rank code indicating a taxon is between genus and species and the\n",
    "   grandparent taxon is at the genus rank.\n",
    "5. NCBI taxonomic ID number\n",
    "6. Indented scientific name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_union_all' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-51-849bb4807bf7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_union_all\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_union_all\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Rank\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'R1'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'df_union_all' is not defined"
     ]
    }
   ],
   "source": [
    "df_union_all.loc[df_union_all[\"Rank\"] == 'R1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.669982416047146"
      ]
     },
     "execution_count": 50,
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "3635181/77841428*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_name = \"RampelliS_2015\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "output = join_metadata(df_union_all, dataset_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_output = os.path.join(path, dataset_name + \".csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'//gfs/data/curated_metagenomes_kraken2/RampelliS_2015.csv'"
      ]
     },
     "execution_count": 57,
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "path_to_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(path_to_output, \"w\") as f: \n",
    "    f.write(output.to_csv())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Name</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>U</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>89.48</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>R</td>\n",
       "      <td>root</td>\n",
       "      <td>10.52</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>R1</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>9.99</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>D</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>9.80</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>D1</td>\n",
       "      <td>Terrabacteria group</td>\n",
       "      <td>5.51</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "output.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL_METADATA = \"\"\"\n",
    "select\n",
    "    DATASET_NAME,\n",
    "    STUDY_CONDITION,\n",
    "    BODY_SITE,\n",
    "    SUBJECTID,\n",
    "    NCBI_ACCESSION,\n",
    "    COUNTRY,\n",
    "    DISEASE,\n",
    "    GENDER,\n",
    "    AGE,\n",
    "    AGE_CATEGORY\n",
    "from METADATA\n",
    "where dataset_name = 'RampelliS_2015'\n",
    "\"\"\"\n",
    "df_metadata = con.select(SQL_METADATA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET_NAME</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>NCBI_ACCESSION</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H1</td>\n",
       "      <td>SRR1927149</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>female</td>\n",
       "      <td>43</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H2</td>\n",
       "      <td>SRR1930133</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>37</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H3</td>\n",
       "      <td>SRR1930145</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>22</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H4</td>\n",
       "      <td>SRR1930149</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>24</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H5</td>\n",
       "      <td>SRR1930176</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>female</td>\n",
       "      <td>70</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "df_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_union_all' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-61-9a3667044cc9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_union_all\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"DATASET_NAME\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_union_all\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Dataset\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mmerged\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_union_all\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_metadata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mon\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'DATASET_NAME'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df_union_all' is not defined"
     ]
    }
   ],
   "source": [
    "df_union_all[\"DATASET_NAME\"] = df_union_all[\"Dataset\"]\n",
    "merged = df_union_all.merge(df_metadata, on='DATASET_NAME')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_to_idx = {}\n",
    "for i in range(len(merged.columns)):\n",
    "    col_to_idx[merged.columns[i]] = i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined = merged.loc[[row[1][col_to_idx[\"Sample_ID\"]] in row[1][col_to_idx[\"NCBI_ACCESSION\"]] for row in merged.iterrows()]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abundance</th>\n",
       "      <th>Root_count</th>\n",
       "      <th>Taxon_count</th>\n",
       "      <th>Rank</th>\n",
       "      <th>NCBI_ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>DATASET_NAME</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>NCBI_ACCESSION</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>89.48</td>\n",
       "      <td>33178082</td>\n",
       "      <td>33178082</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>10.52</td>\n",
       "      <td>3901358</td>\n",
       "      <td>165809</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>root</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>9.99</td>\n",
       "      <td>3704939</td>\n",
       "      <td>3522</td>\n",
       "      <td>R1</td>\n",
       "      <td>131567</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>9.80</td>\n",
       "      <td>3635181</td>\n",
       "      <td>178284</td>\n",
       "      <td>D</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>5.51</td>\n",
       "      <td>2041509</td>\n",
       "      <td>33524</td>\n",
       "      <td>D1</td>\n",
       "      <td>1783272</td>\n",
       "      <td>Terrabacteria group</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "joined.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "minta, projekt, taxonomy_level, baci, abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "output = joined[[\"Sample_ID\", \"Dataset\", \"Rank\", \"Name\", \"Abundance\", \"STUDY_CONDITION\", \"BODY_SITE\", \"SUBJECTID\",\n",
    "                 \"COUNTRY\", \"DISEASE\", \"AGE\", \"AGE_CATEGORY\"]]\n",
    "output = output.reset_index()\n",
    "output = output.drop(columns=\"index\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Name</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>U</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>89.48</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>R</td>\n",
       "      <td>root</td>\n",
       "      <td>10.52</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>R1</td>\n",
       "      <td>cellular organisms</td>\n",
       "      <td>9.99</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>D</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>9.80</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR1930123</td>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>D1</td>\n",
       "      <td>Terrabacteria group</td>\n",
       "      <td>5.51</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H17</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>23</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "output.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata_distinct = pd.DataFrame(df_metadata.NCBI_ACCESSION.str.split(';').tolist(), \n",
    "                                        index = df_metadata.index).stack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0  0    SRR1927149\n",
       "1  0    SRR1930133\n",
       "2  0    SRR1930145\n",
       "3  0    SRR1930149\n",
       "4  0    SRR1930176\n",
       "dtype: object"
      ]
     },
     "execution_count": 65,
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "df_metadata_distinct.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET_NAME</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>NCBI_ACCESSION</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H1</td>\n",
       "      <td>SRR1927149</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>female</td>\n",
       "      <td>43</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H2</td>\n",
       "      <td>SRR1930133</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>37</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H3</td>\n",
       "      <td>SRR1930145</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>22</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H4</td>\n",
       "      <td>SRR1930149</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>24</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H5</td>\n",
       "      <td>SRR1930176</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>female</td>\n",
       "      <td>70</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "df_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL_METADATA = \"\"\"\n",
    "select\n",
    "    DATASET_NAME,\n",
    "    STUDY_CONDITION,\n",
    "    BODY_SITE,\n",
    "    SUBJECTID,\n",
    "    NCBI_ACCESSION,\n",
    "    COUNTRY,\n",
    "    DISEASE,\n",
    "    GENDER,\n",
    "    AGE,\n",
    "    AGE_CATEGORY\n",
    "from METADATA\n",
    "where dataset_name = 'QinJ_2012'\n",
    "\"\"\"\n",
    "df_metadata = con.select(SQL_METADATA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET_NAME</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>NCBI_ACCESSION</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H1</td>\n",
       "      <td>SRR1927149</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>female</td>\n",
       "      <td>43</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H2</td>\n",
       "      <td>SRR1930133</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>37</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H3</td>\n",
       "      <td>SRR1930145</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>22</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H4</td>\n",
       "      <td>SRR1930149</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>24</td>\n",
       "      <td>adult</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RampelliS_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>H5</td>\n",
       "      <td>SRR1930176</td>\n",
       "      <td>TZA</td>\n",
       "      <td>none</td>\n",
       "      <td>female</td>\n",
       "      <td>70</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "df_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata = df_metadata.loc[[x is not None for x in df_metadata.NCBI_ACCESSION]]\n",
    "df_metadata_distinct = pd.DataFrame(df_metadata.NCBI_ACCESSION.str.split(';').tolist(),\n",
    "                                    index = [df_metadata.DATASET_NAME,\n",
    "                                             df_metadata.STUDY_CONDITION, df_metadata.BODY_SITE,\n",
    "                                            df_metadata.SUBJECTID, df_metadata.COUNTRY,\n",
    "                                            df_metadata.DISEASE, df_metadata.GENDER, \n",
    "                                            df_metadata.AGE, df_metadata.AGE_CATEGORY]).stack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DATASET_NAME  STUDY_CONDITION  BODY_SITE  SUBJECTID  COUNTRY  DISEASE  GENDER  AGE   AGE_CATEGORY   \n",
       "QinJ_2012     T2D              stool      DLF001     CHN      T2D      female  59.0  adult         0    SRR341581\n",
       "                                          DLF002     CHN      T2D      female  43.0  adult         0    SRR341582\n",
       "                                          DLF003     CHN      T2D      female  46.0  adult         0    SRR341583\n",
       "                                          DLF004     CHN      T2D      female  25.0  adult         0    SRR341584\n",
       "                                          DLF005     CHN      T2D      female  60.0  adult         0    SRR341585\n",
       "dtype: object"
      ]
     },
     "execution_count": 48,
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "df_metadata_distinct.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = df_metadata_distinct.reset_index([0, 'DATASET_NAME', 'STUDY_CONDITION',\n",
    "    'BODY_SITE',\n",
    "    'SUBJECTID',\n",
    "    'COUNTRY',\n",
    "    'DISEASE',\n",
    "    'GENDER',\n",
    "    'AGE',\n",
    "    'AGE_CATEGORY'])\n",
    "new_df['NCBI_ACCESSION'] = new_df[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = new_df.drop(columns=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET_NAME</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "      <th>NCBI_ACCESSION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BritoIL_2016</td>\n",
       "      <td>control</td>\n",
       "      <td>oralcavity</td>\n",
       "      <td>M1.6</td>\n",
       "      <td>FJI</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>6</td>\n",
       "      <td>child</td>\n",
       "      <td>SRR2231781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BritoIL_2016</td>\n",
       "      <td>control</td>\n",
       "      <td>oralcavity</td>\n",
       "      <td>M1.6</td>\n",
       "      <td>FJI</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>6</td>\n",
       "      <td>child</td>\n",
       "      <td>SRR2243536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BritoIL_2016</td>\n",
       "      <td>control</td>\n",
       "      <td>oralcavity</td>\n",
       "      <td>M1.6</td>\n",
       "      <td>FJI</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>6</td>\n",
       "      <td>child</td>\n",
       "      <td>SRR2243582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BritoIL_2016</td>\n",
       "      <td>control</td>\n",
       "      <td>oralcavity</td>\n",
       "      <td>M1.6</td>\n",
       "      <td>FJI</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>6</td>\n",
       "      <td>child</td>\n",
       "      <td>SRR2243651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BritoIL_2016</td>\n",
       "      <td>control</td>\n",
       "      <td>oralcavity</td>\n",
       "      <td>M1.6</td>\n",
       "      <td>FJI</td>\n",
       "      <td>none</td>\n",
       "      <td>male</td>\n",
       "      <td>6</td>\n",
       "      <td>child</td>\n",
       "      <td>SRR2244452</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "output_type": "execute_result",
     "metadata": {}
    }
   ],
   "source": [
    "new_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"DATASET_NAME\"] = df[\"Dataset\"]\n",
    "df[\"NCBI_ACCESSION\"] = df[\"SampleId\"]\n",
    "\n",
    "merged = df.merge(df_metadata, on='NCBI_ACCESSION')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata_distinct = pd.DataFrame(df_metadata.NCBI_ACCESSION.str.split(';').tolist(),\n",
    "                                    index = [df_metadata.DATASET_NAME,\n",
    "                                             df_metadata.STUDY_CONDITION, df_metadata.BODY_SITE,\n",
    "                                            df_metadata.SUBJECTID, df_metadata.COUNTRY,\n",
    "                                            df_metadata.DISEASE, df_metadata.GENDER,\n",
    "                                            df_metadata.AGE, df_metadata.AGE_CATEGORY]).stack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3.0
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}