{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import seaborn as sns\n",
    "from skbio.diversity import alpha_diversity\n",
    "from skbio.diversity import beta_diversity\n",
    "from skbio import TreeNode\n",
    "from io import StringIO\n",
    "from skbio.stats.ordination import pcoa\n",
    "\n",
    "import cx_Oracle\n",
    "\n",
    "SERVER = 'dboracle.itk.ppke.hu'\n",
    "PORT = 1521\n",
    "SERVICE = 'rsc.itk.ppke.hu'\n",
    "USER = 'metagenome'\n",
    "PASSWORD = 'LrJPRUS73r84'\n",
    "\n",
    "class Connection():\n",
    "\n",
    "    def __init__(self):\n",
    "        dsn_tns = cx_Oracle.makedsn(SERVER, PORT, service_name=SERVICE)\n",
    "        self.connection = cx_Oracle.connect(USER, PASSWORD, dsn_tns)\n",
    "\n",
    "    def get_connection(self):\n",
    "        return self.connection\n",
    "\n",
    "    def select(self, sql):\n",
    "        return pd.read_sql(sql, con=self.connection)\n",
    "\n",
    "con = Connection()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '//gfs/data/curated_metagenomes_metaphlan2/'\n",
    "\n",
    "files = {}\n",
    "# r=root, d=directories, f = files\n",
    "for r, d, f in os.walk(path):\n",
    "    for directory in d:\n",
    "        for r1, d1, f1 in os.walk(os.path.join(path,directory)):\n",
    "            for file in f1:\n",
    "                if '.profile.txt' in file:\n",
    "                    if directory in files:\n",
    "                        files[directory].append(os.path.join(r1, file))\n",
    "                    else:\n",
    "                        files[directory] = [(os.path.join(r1, file))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_data_for_sample(path_to_file):\n",
    "    dataset = path_to_file.split('/')[-2]\n",
    "    sample_id = path_to_file.split('/')[-1].split('.')[0]\n",
    "    \n",
    "    df = pd.read_csv(path_to_file, delimiter = '\\t')\n",
    "    \n",
    "    profile_df_2 = df\n",
    "    profile_df_2['Rank'] = None\n",
    "    profile_df_2['K'] = None\n",
    "    profile_df_2['P'] = None\n",
    "    profile_df_2['C'] = None\n",
    "    profile_df_2['O'] = None\n",
    "    profile_df_2['F'] = None\n",
    "    profile_df_2['G'] = None\n",
    "    profile_df_2['S'] = None\n",
    "    profile_df_2['T'] = None\n",
    "    profile_df_2['Abundance'] = profile_df_2[\"Metaphlan2_Analysis\"]\n",
    "    \n",
    "    for p in profile_df_2[\"#SampleID\"]:\n",
    "        row_level = None\n",
    "        tags = p.split('|')\n",
    "        for tag in tags:\n",
    "            tmp = tag.split('__')\n",
    "            try:\n",
    "                level = tmp[0]\n",
    "                name = tmp[1]\n",
    "                profile_df_2\n",
    "                row_level = level\n",
    "\n",
    "                profile_df_2.loc[profile_df_2['#SampleID'] == p, level] = name\n",
    "            except: \n",
    "                print(\"Can not split:\", tmp)\n",
    "        profile_df_2.loc[profile_df_2['#SampleID'] == p, 'Rank'] = row_level\n",
    "    \n",
    "    def get_name_for_row(row):\n",
    "        try:\n",
    "            return row[row.Rank]\n",
    "        except:\n",
    "            return '???'\n",
    "    \n",
    "    profile_df_2[\"Name\"] = profile_df_2.apply(get_name_for_row, axis=1)\n",
    "\n",
    "    selected = profile_df_2[[\"Rank\", \"Name\", \"Abundance\"]]\n",
    "    \n",
    "    selected['Sample_ID'] = sample_id\n",
    "    selected['Dataset'] = dataset\n",
    "\n",
    "    \n",
    "    return selected[[\"Dataset\", \"Sample_ID\", \"Rank\", \"Name\", \"Abundance\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def join_metadata(df, dataset_name):\n",
    "    SQL_METADATA = \"\"\"\n",
    "        select\n",
    "            DATASET_NAME,\n",
    "            STUDY_CONDITION,\n",
    "            BODY_SITE,\n",
    "            SUBJECTID,\n",
    "            NCBI_ACCESSION,\n",
    "            COUNTRY,\n",
    "            DISEASE,\n",
    "            GENDER,\n",
    "            AGE,\n",
    "            AGE_CATEGORY\n",
    "        from METADATA\n",
    "        where dataset_name = '{dataset_name}'\n",
    "        \"\"\".replace(\"{dataset_name}\", dataset_name)\n",
    "    df_metadata = con.select(SQL_METADATA)\n",
    "    \n",
    "    df[\"DATASET_NAME\"] = df[\"Dataset\"]\n",
    "    merged = df.merge(df_metadata, on='DATASET_NAME')\n",
    "    \n",
    "    col_to_idx = {}\n",
    "    for i in range(len(merged.columns)):\n",
    "        col_to_idx[merged.columns[i]] = i\n",
    "    \n",
    "    joined = merged.loc[[row[1][col_to_idx[\"Sample_ID\"]] in row[1][col_to_idx[\"NCBI_ACCESSION\"]] \n",
    "                        for row in merged.iterrows()]]\n",
    "    \n",
    "    output = joined[[\"Sample_ID\", \"Dataset\", \"Rank\", \"Name\", \"Abundance\", \n",
    "                     \"STUDY_CONDITION\", \"BODY_SITE\", \"SUBJECTID\",\n",
    "                     \"COUNTRY\", \"DISEASE\", \"AGE\", \"AGE_CATEGORY\"]]\n",
    "    output = output.reset_index()\n",
    "    output = output.drop(columns=\"index\")\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AsnicarF_2017  kész.\n",
      "ChngKR_2016  kész.\n",
      "FengQ_2015  kész.\n",
      "Heitz-BuschartA_2016  kész.\n",
      "KarlssonFH_2013  kész.\n",
      "LeChatelierE_2013  kész.\n",
      "LiuW_2016  kész.\n",
      "LomanNJ_2013  kész.\n",
      "Obregon-TitoAJ_2015  kész.\n",
      "QinJ_2012  kész.\n",
      "QinN_2014  kész.\n",
      "RampelliS_2015  kész.\n",
      "SchirmerM_2016  kész.\n",
      "VatanenT_2016  kész.\n",
      "VogtmannE_2016  kész.\n",
      "XieH_2016  kész.\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018191.profile.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gfs/progs/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/gfs/progs/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018205.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018223.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018198.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018212.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018199.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018206.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018213.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018192.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018185.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018200.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018214.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018207.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018193.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018186.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018201.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018208.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018215.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018187.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018194.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018202.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018209.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018188.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018195.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018216.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018203.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018210.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018189.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018217.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018196.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018204.profile.txt\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018211.profile.txt\n"
     ]
    }
   ],
   "source": [
    "dataset_dfs = {}\n",
    "i = 0;\n",
    "for dataset_name, filelist in files.items():\n",
    "    if dataset_name == \"YuJ_2015\":\n",
    "        df_list = []\n",
    "        fc = 0\n",
    "        for file in filelist:\n",
    "            print(file)\n",
    "            df = get_data_for_sample(file)\n",
    "            if df is not None:\n",
    "                df_list.append(get_data_for_sample(file))\n",
    "            fc = fc+1\n",
    "        df_dataset = pd.concat(df_list)\n",
    "        output = join_metadata(df_dataset, dataset_name)\n",
    "        print(\"Files in dataset:\", fc)\n",
    "        \n",
    "        path_to_output = os.path.join(path, dataset_name + \".csv\")\n",
    "        with open(path_to_output, \"w\") as f: \n",
    "            f.write(output.to_csv()) \n",
    "    i = i+1\n",
    "    print(dataset_name, \" kész.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "128"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(files['YuJ_2015'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_file = files[\"AsnicarF_2017\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'//gfs/data/curated_metagenomes_metaphlan2/AsnicarF_2017/SRR4052021.profile.txt'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_to_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       #SampleID  Metaphlan2_Analysis Rank         k  \\\n",
      "0                    k__Bacteria             97.92389    k  Bacteria   \n",
      "1                     k__Viruses              2.07611    k   Viruses   \n",
      "2  k__Bacteria|p__Proteobacteria             78.59277    p  Bacteria   \n",
      "3  k__Bacteria|p__Actinobacteria             12.60032    p  Bacteria   \n",
      "4      k__Bacteria|p__Firmicutes              6.73081    p  Bacteria   \n",
      "\n",
      "                p     c     o     f     g     s     t  Abundance  \n",
      "0            None  None  None  None  None  None  None   97.92389  \n",
      "1            None  None  None  None  None  None  None    2.07611  \n",
      "2  Proteobacteria  None  None  None  None  None  None   78.59277  \n",
      "3  Actinobacteria  None  None  None  None  None  None   12.60032  \n",
      "4      Firmicutes  None  None  None  None  None  None    6.73081  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gfs/progs/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/gfs/progs/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:38: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "df = get_data_for_sample(path_to_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Name</th>\n",
       "      <th>Abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>97.92389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>k</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>2.07611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>p</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>78.59277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>p</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>12.60032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>p</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>6.73081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Dataset   Sample_ID Rank            Name  Abundance\n",
       "0  AsnicarF_2017  SRR4052021    k        Bacteria   97.92389\n",
       "1  AsnicarF_2017  SRR4052021    k         Viruses    2.07611\n",
       "2  AsnicarF_2017  SRR4052021    p  Proteobacteria   78.59277\n",
       "3  AsnicarF_2017  SRR4052021    p  Actinobacteria   12.60032\n",
       "4  AsnicarF_2017  SRR4052021    p      Firmicutes    6.73081"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined = join_metadata(df, \"AsnicarF_2017\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Name</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>97.92389</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>MV_FEI1</td>\n",
       "      <td>ITA</td>\n",
       "      <td>healthy</td>\n",
       "      <td>1.0</td>\n",
       "      <td>newborn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>k</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>2.07611</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>MV_FEI1</td>\n",
       "      <td>ITA</td>\n",
       "      <td>healthy</td>\n",
       "      <td>1.0</td>\n",
       "      <td>newborn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>p</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>78.59277</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>MV_FEI1</td>\n",
       "      <td>ITA</td>\n",
       "      <td>healthy</td>\n",
       "      <td>1.0</td>\n",
       "      <td>newborn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>p</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>12.60032</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>MV_FEI1</td>\n",
       "      <td>ITA</td>\n",
       "      <td>healthy</td>\n",
       "      <td>1.0</td>\n",
       "      <td>newborn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR4052021</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>p</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>6.73081</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>MV_FEI1</td>\n",
       "      <td>ITA</td>\n",
       "      <td>healthy</td>\n",
       "      <td>1.0</td>\n",
       "      <td>newborn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample_ID        Dataset Rank            Name  Abundance STUDY_CONDITION  \\\n",
       "0  SRR4052021  AsnicarF_2017    k        Bacteria   97.92389         control   \n",
       "1  SRR4052021  AsnicarF_2017    k         Viruses    2.07611         control   \n",
       "2  SRR4052021  AsnicarF_2017    p  Proteobacteria   78.59277         control   \n",
       "3  SRR4052021  AsnicarF_2017    p  Actinobacteria   12.60032         control   \n",
       "4  SRR4052021  AsnicarF_2017    p      Firmicutes    6.73081         control   \n",
       "\n",
       "  BODY_SITE SUBJECTID COUNTRY  DISEASE  AGE AGE_CATEGORY  \n",
       "0     stool   MV_FEI1     ITA  healthy  1.0      newborn  \n",
       "1     stool   MV_FEI1     ITA  healthy  1.0      newborn  \n",
       "2     stool   MV_FEI1     ITA  healthy  1.0      newborn  \n",
       "3     stool   MV_FEI1     ITA  healthy  1.0      newborn  \n",
       "4     stool   MV_FEI1     ITA  healthy  1.0      newborn  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "profile_df = pd.read_csv(path_to_profile, delimiter = '\\t', skiprows = [1])\n",
    "\n",
    "#f = open(\"/gfs/data/curated_metagenomes_kraken2/RampelliS_2015⁩/profiled_metagenome.txt\", \n",
    "#encoding=\"utf-8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['AsnicarF_2017', 'ChngKR_2016', 'FengQ_2015', 'Heitz-BuschartA_2016', 'KarlssonFH_2013', 'LeChatelierE_2013', 'LiuW_2016', 'LomanNJ_2013', 'Obregon-TitoAJ_2015', 'QinJ_2012', 'QinN_2014', 'RampelliS_2015', 'SchirmerM_2016', 'VatanenT_2016', 'VogtmannE_2016', 'XieH_2016', 'YuJ_2015', 'ZellerG_2014'])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_file = files[\"AsnicarF_2017\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = path_to_file.split('/')[-2]\n",
    "df = pd.read_csv(path_to_file, delimiter = '\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>Metaphlan2_Analysis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>97.92389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Viruses</td>\n",
       "      <td>2.07611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria|p__Proteobacteria</td>\n",
       "      <td>78.59277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria</td>\n",
       "      <td>12.60032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria|p__Firmicutes</td>\n",
       "      <td>6.73081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       #SampleID  Metaphlan2_Analysis\n",
       "0                    k__Bacteria             97.92389\n",
       "1                     k__Viruses              2.07611\n",
       "2  k__Bacteria|p__Proteobacteria             78.59277\n",
       "3  k__Bacteria|p__Actinobacteria             12.60032\n",
       "4      k__Bacteria|p__Firmicutes              6.73081"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "profile_df_2 = df\n",
    "profile_df_2['level'] = None\n",
    "profile_df_2['k'] = None\n",
    "profile_df_2['p'] = None\n",
    "profile_df_2['c'] = None\n",
    "profile_df_2['o'] = None\n",
    "profile_df_2['f'] = None\n",
    "profile_df_2['g'] = None\n",
    "profile_df_2['s'] = None\n",
    "profile_df_2['t'] = None\n",
    "profile_df_2['value'] = profile_df_2[\"Metaphlan2_Analysis\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>Metaphlan2_Analysis</th>\n",
       "      <th>level</th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "      <th>t</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>97.92389</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>97.92389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Viruses</td>\n",
       "      <td>2.07611</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2.07611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria|p__Proteobacteria</td>\n",
       "      <td>78.59277</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>78.59277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria</td>\n",
       "      <td>12.60032</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>12.60032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria|p__Firmicutes</td>\n",
       "      <td>6.73081</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6.73081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       #SampleID  Metaphlan2_Analysis level     k     p     c  \\\n",
       "0                    k__Bacteria             97.92389  None  None  None  None   \n",
       "1                     k__Viruses              2.07611  None  None  None  None   \n",
       "2  k__Bacteria|p__Proteobacteria             78.59277  None  None  None  None   \n",
       "3  k__Bacteria|p__Actinobacteria             12.60032  None  None  None  None   \n",
       "4      k__Bacteria|p__Firmicutes              6.73081  None  None  None  None   \n",
       "\n",
       "      o     f     g     s     t     value  \n",
       "0  None  None  None  None  None  97.92389  \n",
       "1  None  None  None  None  None   2.07611  \n",
       "2  None  None  None  None  None  78.59277  \n",
       "3  None  None  None  None  None  12.60032  \n",
       "4  None  None  None  None  None   6.73081  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "profile_df_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "for p in profile_df_2[\"#SampleID\"]:\n",
    "    row_level = None\n",
    "    tags = p.split('|')\n",
    "    for tag in tags:\n",
    "        tmp = tag.split('__')\n",
    "        level = tmp[0]\n",
    "        name = tmp[1]\n",
    "        profile_df_2\n",
    "        row_level = level\n",
    "        \n",
    "        profile_df_2.loc[profile_df_2['#SampleID'] == p, level] = name\n",
    "    profile_df_2.loc[profile_df_2['#SampleID'] == p, 'level'] = row_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>Metaphlan2_Analysis</th>\n",
       "      <th>level</th>\n",
       "      <th>k</th>\n",
       "      <th>p</th>\n",
       "      <th>c</th>\n",
       "      <th>o</th>\n",
       "      <th>f</th>\n",
       "      <th>g</th>\n",
       "      <th>s</th>\n",
       "      <th>t</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k__Bacteria</td>\n",
       "      <td>97.92389</td>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>97.92389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k__Viruses</td>\n",
       "      <td>2.07611</td>\n",
       "      <td>k</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2.07611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k__Bacteria|p__Proteobacteria</td>\n",
       "      <td>78.59277</td>\n",
       "      <td>p</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>78.59277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k__Bacteria|p__Actinobacteria</td>\n",
       "      <td>12.60032</td>\n",
       "      <td>p</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>12.60032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k__Bacteria|p__Firmicutes</td>\n",
       "      <td>6.73081</td>\n",
       "      <td>p</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>6.73081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       #SampleID  Metaphlan2_Analysis level         k  \\\n",
       "0                    k__Bacteria             97.92389     k  Bacteria   \n",
       "1                     k__Viruses              2.07611     k   Viruses   \n",
       "2  k__Bacteria|p__Proteobacteria             78.59277     p  Bacteria   \n",
       "3  k__Bacteria|p__Actinobacteria             12.60032     p  Bacteria   \n",
       "4      k__Bacteria|p__Firmicutes              6.73081     p  Bacteria   \n",
       "\n",
       "                p     c     o     f     g     s     t     value  \n",
       "0            None  None  None  None  None  None  None  97.92389  \n",
       "1            None  None  None  None  None  None  None   2.07611  \n",
       "2  Proteobacteria  None  None  None  None  None  None  78.59277  \n",
       "3  Actinobacteria  None  None  None  None  None  None  12.60032  \n",
       "4      Firmicutes  None  None  None  None  None  None   6.73081  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "profile_df_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "profile_df_2[\"Name\"] = [profile_df_2[i] for i in profile_df_2[\"level\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "profile_df_2[\"Name\"]  = profile_df_2.apply(lambda row: row[row.level], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected = profile_df_2[[\"level\", \"Name\", \"value\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level</th>\n",
       "      <th>Name</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>97.92389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>2.07611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>p</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>78.59277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>p</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>12.60032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>p</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>6.73081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  level            Name     value\n",
       "0     k        Bacteria  97.92389\n",
       "1     k         Viruses   2.07611\n",
       "2     p  Proteobacteria  78.59277\n",
       "3     p  Actinobacteria  12.60032\n",
       "4     p      Firmicutes   6.73081"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "b = pd.read_csv('/gfs/data/curated_metagenomes_metaphlan2/ChngKR_2016/SRR1950687.bowtie2out.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1316718, 1)"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4816706"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(2407746+2408960)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27.33648265017628"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1316718/4816706*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b = pd.read_csv('/gfs/data/curated_metagenomes_metaphlan2/ChngKR_2016/SRR1950687.bowtie2out.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "ename": "ParserError",
     "evalue": "Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mParserError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-105-a69a7c46df16>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/gfs/data/curated_metagenomes_preprocessed/ChngKR_2016/trimmed_files/SRR1950687_1.fastq.trimmed.gz'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/gfs/progs/miniconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, doublequote, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    676\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    677\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 678\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    679\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    680\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gfs/progs/miniconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    444\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    445\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 446\u001b[0;31m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    447\u001b[0m     \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    448\u001b[0m         \u001b[0mparser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gfs/progs/miniconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1034\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'skipfooter not supported for iteration'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1035\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1036\u001b[0;31m         \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1037\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1038\u001b[0m         \u001b[0;31m# May alter columns / col_dict\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gfs/progs/miniconda3/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1846\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnrows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1847\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1848\u001b[0;31m             \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1849\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1850\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_first_chunk\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.read\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._read_low_memory\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._read_rows\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._tokenize_rows\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.raise_parser_error\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mParserError\u001b[0m: Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'."
     ]
    }
   ],
   "source": [
    "p = pd.read_csv('/gfs/data/curated_metagenomes_preprocessed/ChngKR_2016/trimmed_files/SRR1950687_1.fastq.trimmed.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AsnicarF_2017  kész.\n",
      "ChngKR_2016  kész.\n",
      "FengQ_2015  kész.\n",
      "Heitz-BuschartA_2016  kész.\n",
      "KarlssonFH_2013  kész.\n",
      "LeChatelierE_2013  kész.\n",
      "LiuW_2016  kész.\n",
      "LomanNJ_2013  kész.\n",
      "Obregon-TitoAJ_2015  kész.\n",
      "QinJ_2012  kész.\n",
      "QinN_2014  kész.\n",
      "RampelliS_2015  kész.\n",
      "SchirmerM_2016  kész.\n",
      "VatanenT_2016  kész.\n",
      "VogtmannE_2016  kész.\n",
      "XieH_2016  kész.\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018191.profile.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gfs/progs/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:45: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/gfs/progs/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018191    k        Bacteria   99.98996\n",
      "1  YuJ_2015  ERR1018191    k         Viruses    0.01004\n",
      "2  YuJ_2015  ERR1018191    p      Firmicutes   58.59706\n",
      "3  YuJ_2015  ERR1018191    p  Actinobacteria   18.78107\n",
      "4  YuJ_2015  ERR1018191    p  Proteobacteria   18.20503\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018205.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018205    k        Bacteria   99.91806\n",
      "1  YuJ_2015  ERR1018205    k       Eukaryota    0.08194\n",
      "2  YuJ_2015  ERR1018205    p      Firmicutes   41.98932\n",
      "3  YuJ_2015  ERR1018205    p   Bacteroidetes   36.04325\n",
      "4  YuJ_2015  ERR1018205    p  Actinobacteria   17.76978\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018223.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018223    k        Bacteria   99.60468\n",
      "1  YuJ_2015  ERR1018223    k         Viruses    0.39532\n",
      "2  YuJ_2015  ERR1018223    p      Firmicutes   46.17671\n",
      "3  YuJ_2015  ERR1018223    p   Bacteroidetes   33.23399\n",
      "4  YuJ_2015  ERR1018223    p  Actinobacteria   12.15795\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018198.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018198    k        Bacteria   99.69974\n",
      "1  YuJ_2015  ERR1018198    k         Viruses    0.30026\n",
      "2  YuJ_2015  ERR1018198    p      Firmicutes   68.81646\n",
      "3  YuJ_2015  ERR1018198    p   Bacteroidetes   15.99558\n",
      "4  YuJ_2015  ERR1018198    p  Proteobacteria    9.50850\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018212.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018212    k        Bacteria   99.97517\n",
      "1  YuJ_2015  ERR1018212    k         Archaea    0.02483\n",
      "2  YuJ_2015  ERR1018212    p      Firmicutes   47.04173\n",
      "3  YuJ_2015  ERR1018212    p  Actinobacteria   38.83921\n",
      "4  YuJ_2015  ERR1018212    p   Bacteroidetes   11.40266\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018199.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018199    k        Bacteria   99.92959\n",
      "1  YuJ_2015  ERR1018199    k         Viruses    0.07041\n",
      "2  YuJ_2015  ERR1018199    p      Firmicutes   57.01098\n",
      "3  YuJ_2015  ERR1018199    p   Bacteroidetes   35.15596\n",
      "4  YuJ_2015  ERR1018199    p  Actinobacteria    6.50197\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018206.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018206    k        Bacteria   99.97673\n",
      "1  YuJ_2015  ERR1018206    k         Viruses    0.02327\n",
      "2  YuJ_2015  ERR1018206    p   Bacteroidetes   47.64421\n",
      "3  YuJ_2015  ERR1018206    p      Firmicutes   31.35981\n",
      "4  YuJ_2015  ERR1018206    p  Actinobacteria   20.02351\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018213.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018213    k        Bacteria   99.84561\n",
      "1  YuJ_2015  ERR1018213    k         Viruses    0.14993\n",
      "2  YuJ_2015  ERR1018213    k       Eukaryota    0.00445\n",
      "3  YuJ_2015  ERR1018213    p      Firmicutes   79.58625\n",
      "4  YuJ_2015  ERR1018213    p  Actinobacteria    9.92787\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018192.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018192    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018192    p      Firmicutes   55.47020\n",
      "2  YuJ_2015  ERR1018192    p   Bacteroidetes   26.11597\n",
      "3  YuJ_2015  ERR1018192    p  Actinobacteria   12.85577\n",
      "4  YuJ_2015  ERR1018192    p  Proteobacteria    5.19836\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018185.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018185    k         Bacteria   98.68889\n",
      "1  YuJ_2015  ERR1018185    k          Archaea    0.74715\n",
      "2  YuJ_2015  ERR1018185    k          Viruses    0.56395\n",
      "3  YuJ_2015  ERR1018185    p       Firmicutes   32.89464\n",
      "4  YuJ_2015  ERR1018185    p  Verrucomicrobia   30.09550\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018200.profile.txt\n",
      "    Dataset   Sample_ID Rank        Name  Abundance\n",
      "0  YuJ_2015  ERR1018200    k    Bacteria   99.96873\n",
      "1  YuJ_2015  ERR1018200    k     Archaea    0.01784\n",
      "2  YuJ_2015  ERR1018200    k     Viruses    0.01013\n",
      "3  YuJ_2015  ERR1018200    k   Eukaryota    0.00329\n",
      "4  YuJ_2015  ERR1018200    p  Firmicutes   64.33837\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018214.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018214    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018214    p   Bacteroidetes   60.14595\n",
      "2  YuJ_2015  ERR1018214    p      Firmicutes   36.64468\n",
      "3  YuJ_2015  ERR1018214    p  Proteobacteria    1.68630\n",
      "4  YuJ_2015  ERR1018214    p    Fusobacteria    1.05996\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018207.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018207    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018207    p   Bacteroidetes   56.28066\n",
      "2  YuJ_2015  ERR1018207    p      Firmicutes   37.80477\n",
      "3  YuJ_2015  ERR1018207    p  Proteobacteria    3.81560\n",
      "4  YuJ_2015  ERR1018207    p  Actinobacteria    1.93717\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018193.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018193    k        Bacteria   99.99366\n",
      "1  YuJ_2015  ERR1018193    k         Archaea    0.00634\n",
      "2  YuJ_2015  ERR1018193    p   Bacteroidetes   45.48090\n",
      "3  YuJ_2015  ERR1018193    p  Proteobacteria   30.16261\n",
      "4  YuJ_2015  ERR1018193    p      Firmicutes   23.41289\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018186.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018186    k         Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018186    p       Firmicutes   39.95370\n",
      "2  YuJ_2015  ERR1018186    p    Bacteroidetes   32.94505\n",
      "3  YuJ_2015  ERR1018186    p  Verrucomicrobia   17.28948\n",
      "4  YuJ_2015  ERR1018186    p   Proteobacteria    7.09903\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018201.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018201    k        Bacteria   76.91057\n",
      "1  YuJ_2015  ERR1018201    k         Viruses   23.08943\n",
      "2  YuJ_2015  ERR1018201    p   Bacteroidetes   49.94147\n",
      "3  YuJ_2015  ERR1018201    p  Viruses_noname   23.08943\n",
      "4  YuJ_2015  ERR1018201    p      Firmicutes   18.64759\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018208.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018208    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018208    p  Proteobacteria   86.07760\n",
      "2  YuJ_2015  ERR1018208    p      Firmicutes   11.59370\n",
      "3  YuJ_2015  ERR1018208    p   Bacteroidetes    1.17934\n",
      "4  YuJ_2015  ERR1018208    p  Actinobacteria    1.12391\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018215.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018215    k       Bacteria   99.96724\n",
      "1  YuJ_2015  ERR1018215    k      Eukaryota    0.02965\n",
      "2  YuJ_2015  ERR1018215    k        Viruses    0.00311\n",
      "3  YuJ_2015  ERR1018215    p     Firmicutes   64.41147\n",
      "4  YuJ_2015  ERR1018215    p  Bacteroidetes   18.38415\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018187.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018187    k        Bacteria   75.14071\n",
      "1  YuJ_2015  ERR1018187    k         Archaea   24.78691\n",
      "2  YuJ_2015  ERR1018187    k         Viruses    0.07239\n",
      "3  YuJ_2015  ERR1018187    p  Actinobacteria   47.09311\n",
      "4  YuJ_2015  ERR1018187    p   Euryarchaeota   24.78691\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018194.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018194    k       Bacteria   98.86351\n",
      "1  YuJ_2015  ERR1018194    k        Viruses    1.13601\n",
      "2  YuJ_2015  ERR1018194    k        Archaea    0.00049\n",
      "3  YuJ_2015  ERR1018194    p  Bacteroidetes   45.40390\n",
      "4  YuJ_2015  ERR1018194    p     Firmicutes   27.49264\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018202.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018202    k         Viruses   53.15019\n",
      "1  YuJ_2015  ERR1018202    k        Bacteria   46.82947\n",
      "2  YuJ_2015  ERR1018202    k         Archaea    0.02034\n",
      "3  YuJ_2015  ERR1018202    p  Viruses_noname   53.15019\n",
      "4  YuJ_2015  ERR1018202    p  Proteobacteria   21.95771\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018209.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018209    k       Bacteria   99.98731\n",
      "1  YuJ_2015  ERR1018209    k        Viruses    0.00635\n",
      "2  YuJ_2015  ERR1018209    k      Eukaryota    0.00634\n",
      "3  YuJ_2015  ERR1018209    p  Bacteroidetes   52.05363\n",
      "4  YuJ_2015  ERR1018209    p     Firmicutes   37.55983\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018188.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018188    k         Bacteria   99.99639\n",
      "1  YuJ_2015  ERR1018188    k          Archaea    0.00361\n",
      "2  YuJ_2015  ERR1018188    p       Firmicutes   65.78088\n",
      "3  YuJ_2015  ERR1018188    p    Bacteroidetes   17.37403\n",
      "4  YuJ_2015  ERR1018188    p  Verrucomicrobia    6.83097\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018195.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018195    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018195    p   Bacteroidetes   59.96187\n",
      "2  YuJ_2015  ERR1018195    p      Firmicutes   20.43187\n",
      "3  YuJ_2015  ERR1018195    p  Actinobacteria   16.02082\n",
      "4  YuJ_2015  ERR1018195    p  Proteobacteria    3.57631\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018216.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018216    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018216    p      Firmicutes   50.69890\n",
      "2  YuJ_2015  ERR1018216    p  Actinobacteria   29.95196\n",
      "3  YuJ_2015  ERR1018216    p   Bacteroidetes   16.71789\n",
      "4  YuJ_2015  ERR1018216    p  Proteobacteria    2.51008\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018203.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018203    k         Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018203    p    Bacteroidetes   46.83428\n",
      "2  YuJ_2015  ERR1018203    p       Firmicutes   26.78259\n",
      "3  YuJ_2015  ERR1018203    p     Fusobacteria    9.77037\n",
      "4  YuJ_2015  ERR1018203    p  Verrucomicrobia    6.61656\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018210.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018210    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018210    p   Bacteroidetes   37.68764\n",
      "2  YuJ_2015  ERR1018210    p      Firmicutes   34.70381\n",
      "3  YuJ_2015  ERR1018210    p  Actinobacteria   23.76541\n",
      "4  YuJ_2015  ERR1018210    p  Proteobacteria    2.67726\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018189.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018189    k        Bacteria   99.31011\n",
      "1  YuJ_2015  ERR1018189    k         Viruses    0.68989\n",
      "2  YuJ_2015  ERR1018189    p      Firmicutes   31.13977\n",
      "3  YuJ_2015  ERR1018189    p   Bacteroidetes   30.30669\n",
      "4  YuJ_2015  ERR1018189    p  Actinobacteria   19.81183\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018217.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018217    k        Bacteria   99.82616\n",
      "1  YuJ_2015  ERR1018217    k       Eukaryota    0.17384\n",
      "2  YuJ_2015  ERR1018217    p      Firmicutes   75.57748\n",
      "3  YuJ_2015  ERR1018217    p  Proteobacteria   12.38027\n",
      "4  YuJ_2015  ERR1018217    p  Actinobacteria   10.75241\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018196.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018196    k       Bacteria   95.54313\n",
      "1  YuJ_2015  ERR1018196    k        Archaea    4.45687\n",
      "2  YuJ_2015  ERR1018196    p  Bacteroidetes   51.25023\n",
      "3  YuJ_2015  ERR1018196    p     Firmicutes   40.16185\n",
      "4  YuJ_2015  ERR1018196    p  Euryarchaeota    4.45687\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018204.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018204    k         Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018204    p  Verrucomicrobia   61.21335\n",
      "2  YuJ_2015  ERR1018204    p       Firmicutes   31.35979\n",
      "3  YuJ_2015  ERR1018204    p   Proteobacteria    3.29547\n",
      "4  YuJ_2015  ERR1018204    p    Bacteroidetes    2.73614\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018211.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018211    k       Bacteria   97.16621\n",
      "1  YuJ_2015  ERR1018211    k        Viruses    2.81427\n",
      "2  YuJ_2015  ERR1018211    k        Archaea    0.01952\n",
      "3  YuJ_2015  ERR1018211    p     Firmicutes   52.46616\n",
      "4  YuJ_2015  ERR1018211    p  Bacteroidetes   22.81343\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018190.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018190    k       Bacteria   94.42236\n",
      "1  YuJ_2015  ERR1018190    k        Viruses    5.57624\n",
      "2  YuJ_2015  ERR1018190    k        Archaea    0.00140\n",
      "3  YuJ_2015  ERR1018190    p  Bacteroidetes   80.89686\n",
      "4  YuJ_2015  ERR1018190    p     Firmicutes   10.15024\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018218.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018218    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018218    p      Firmicutes   55.65712\n",
      "2  YuJ_2015  ERR1018218    p   Bacteroidetes   39.69842\n",
      "3  YuJ_2015  ERR1018218    p  Proteobacteria    3.48344\n",
      "4  YuJ_2015  ERR1018218    p  Actinobacteria    1.08437\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018219.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018219    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018219    p   Bacteroidetes   60.11851\n",
      "2  YuJ_2015  ERR1018219    p  Proteobacteria   13.85119\n",
      "3  YuJ_2015  ERR1018219    p      Firmicutes   13.70131\n",
      "4  YuJ_2015  ERR1018219    p    Fusobacteria    8.28268\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018197.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018197    k        Bacteria   99.96014\n",
      "1  YuJ_2015  ERR1018197    k         Archaea    0.03986\n",
      "2  YuJ_2015  ERR1018197    p      Firmicutes   34.88656\n",
      "3  YuJ_2015  ERR1018197    p  Actinobacteria   34.68022\n",
      "4  YuJ_2015  ERR1018197    p   Bacteroidetes   27.78196\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018227.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018227    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018227    p      Firmicutes   61.07940\n",
      "2  YuJ_2015  ERR1018227    p   Bacteroidetes   29.83334\n",
      "3  YuJ_2015  ERR1018227    p    Fusobacteria    4.79178\n",
      "4  YuJ_2015  ERR1018227    p  Proteobacteria    2.78925\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018234.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018234    k        Bacteria   99.99681\n",
      "1  YuJ_2015  ERR1018234    k       Eukaryota    0.00319\n",
      "2  YuJ_2015  ERR1018234    p  Proteobacteria   95.11760\n",
      "3  YuJ_2015  ERR1018234    p      Firmicutes    4.03664\n",
      "4  YuJ_2015  ERR1018234    p  Actinobacteria    0.71727\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018241.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018241    k        Bacteria   99.99123\n",
      "1  YuJ_2015  ERR1018241    k         Viruses    0.00877\n",
      "2  YuJ_2015  ERR1018241    p   Bacteroidetes   66.65117\n",
      "3  YuJ_2015  ERR1018241    p      Firmicutes   23.88123\n",
      "4  YuJ_2015  ERR1018241    p  Proteobacteria    5.00323\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018248.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018248    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018248    p      Firmicutes   58.90822\n",
      "2  YuJ_2015  ERR1018248    p   Bacteroidetes   19.66618\n",
      "3  YuJ_2015  ERR1018248    p  Proteobacteria   13.96020\n",
      "4  YuJ_2015  ERR1018248    p  Actinobacteria    7.45014\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018220.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018220    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018220    p      Firmicutes   60.02574\n",
      "2  YuJ_2015  ERR1018220    p   Bacteroidetes   36.20974\n",
      "3  YuJ_2015  ERR1018220    p  Proteobacteria    2.14869\n",
      "4  YuJ_2015  ERR1018220    p  Actinobacteria    1.60224\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018235.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018235    k       Bacteria   99.19999\n",
      "1  YuJ_2015  ERR1018235    k        Archaea    0.76599\n",
      "2  YuJ_2015  ERR1018235    k        Viruses    0.03402\n",
      "3  YuJ_2015  ERR1018235    p     Firmicutes   69.71990\n",
      "4  YuJ_2015  ERR1018235    p  Bacteroidetes   25.39344\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018228.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018228    k        Bacteria   98.47092\n",
      "1  YuJ_2015  ERR1018228    k         Archaea    1.52908\n",
      "2  YuJ_2015  ERR1018228    p      Firmicutes   71.77452\n",
      "3  YuJ_2015  ERR1018228    p  Actinobacteria   13.52044\n",
      "4  YuJ_2015  ERR1018228    p   Bacteroidetes   13.00964\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018249.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018249    k       Bacteria   77.59802\n",
      "1  YuJ_2015  ERR1018249    k        Viruses   22.39415\n",
      "2  YuJ_2015  ERR1018249    k        Archaea    0.00782\n",
      "3  YuJ_2015  ERR1018249    p  Bacteroidetes   45.47741\n",
      "4  YuJ_2015  ERR1018249    p     Firmicutes   28.69254\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018242.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018242    k        Bacteria   99.99956\n",
      "1  YuJ_2015  ERR1018242    k         Archaea    0.00044\n",
      "2  YuJ_2015  ERR1018242    p      Firmicutes   43.61787\n",
      "3  YuJ_2015  ERR1018242    p  Proteobacteria   41.45010\n",
      "4  YuJ_2015  ERR1018242    p   Bacteroidetes   12.98437\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018221.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018221    k         Bacteria   99.99638\n",
      "1  YuJ_2015  ERR1018221    k          Archaea    0.00362\n",
      "2  YuJ_2015  ERR1018221    p       Firmicutes   54.41279\n",
      "3  YuJ_2015  ERR1018221    p    Bacteroidetes   25.89281\n",
      "4  YuJ_2015  ERR1018221    p  Verrucomicrobia   11.32364\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018229.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018229    k        Bacteria   99.92109\n",
      "1  YuJ_2015  ERR1018229    k         Viruses    0.07525\n",
      "2  YuJ_2015  ERR1018229    k       Eukaryota    0.00365\n",
      "3  YuJ_2015  ERR1018229    p      Firmicutes   85.87620\n",
      "4  YuJ_2015  ERR1018229    p  Actinobacteria   11.26357\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018236.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018236    k         Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018236    p    Bacteroidetes   75.01670\n",
      "2  YuJ_2015  ERR1018236    p       Firmicutes   15.45091\n",
      "3  YuJ_2015  ERR1018236    p  Verrucomicrobia    4.84288\n",
      "4  YuJ_2015  ERR1018236    p   Proteobacteria    3.27159\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018250.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018250    k        Bacteria   99.94850\n",
      "1  YuJ_2015  ERR1018250    k         Archaea    0.05150\n",
      "2  YuJ_2015  ERR1018250    p      Firmicutes   83.65006\n",
      "3  YuJ_2015  ERR1018250    p   Bacteroidetes    9.42719\n",
      "4  YuJ_2015  ERR1018250    p  Actinobacteria    4.58112\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018243.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018243    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018243    p   Bacteroidetes   72.80476\n",
      "2  YuJ_2015  ERR1018243    p  Actinobacteria   21.57522\n",
      "3  YuJ_2015  ERR1018243    p      Firmicutes    3.40132\n",
      "4  YuJ_2015  ERR1018243    p  Proteobacteria    2.17651\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018222.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018222    k         Bacteria   95.74609\n",
      "1  YuJ_2015  ERR1018222    k          Archaea    4.25003\n",
      "2  YuJ_2015  ERR1018222    k          Viruses    0.00387\n",
      "3  YuJ_2015  ERR1018222    p  Verrucomicrobia   36.87922\n",
      "4  YuJ_2015  ERR1018222    p    Bacteroidetes   36.10739\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018251.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018251    k         Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018251    p    Bacteroidetes   45.22364\n",
      "2  YuJ_2015  ERR1018251    p       Firmicutes   39.47050\n",
      "3  YuJ_2015  ERR1018251    p   Proteobacteria   13.23895\n",
      "4  YuJ_2015  ERR1018251    p  Verrucomicrobia    1.37521\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018230.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018230    k         Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018230    p    Bacteroidetes   39.09956\n",
      "2  YuJ_2015  ERR1018230    p       Firmicutes   31.67871\n",
      "3  YuJ_2015  ERR1018230    p  Verrucomicrobia   16.24596\n",
      "4  YuJ_2015  ERR1018230    p   Actinobacteria    9.62532\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018237.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018237    k        Bacteria   99.99720\n",
      "1  YuJ_2015  ERR1018237    k         Viruses    0.00280\n",
      "2  YuJ_2015  ERR1018237    p   Bacteroidetes   55.83232\n",
      "3  YuJ_2015  ERR1018237    p      Firmicutes   24.37030\n",
      "4  YuJ_2015  ERR1018237    p  Proteobacteria   15.69100\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018224.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018224    k       Bacteria   99.95307\n",
      "1  YuJ_2015  ERR1018224    k      Eukaryota    0.04306\n",
      "2  YuJ_2015  ERR1018224    k        Viruses    0.00387\n",
      "3  YuJ_2015  ERR1018224    p     Firmicutes   87.49441\n",
      "4  YuJ_2015  ERR1018224    p  Bacteroidetes    7.53720\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018244.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018244    k        Bacteria   99.98895\n",
      "1  YuJ_2015  ERR1018244    k         Archaea    0.01105\n",
      "2  YuJ_2015  ERR1018244    p   Bacteroidetes   60.08264\n",
      "3  YuJ_2015  ERR1018244    p      Firmicutes   28.38731\n",
      "4  YuJ_2015  ERR1018244    p  Actinobacteria    9.09663\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018252.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018252    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018252    p      Firmicutes   90.62075\n",
      "2  YuJ_2015  ERR1018252    p  Actinobacteria    5.40992\n",
      "3  YuJ_2015  ERR1018252    p   Bacteroidetes    2.96796\n",
      "4  YuJ_2015  ERR1018252    p  Proteobacteria    0.98733\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018231.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018231    k         Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018231    p    Bacteroidetes   56.26013\n",
      "2  YuJ_2015  ERR1018231    p       Firmicutes   34.14358\n",
      "3  YuJ_2015  ERR1018231    p   Proteobacteria    5.24339\n",
      "4  YuJ_2015  ERR1018231    p  Verrucomicrobia    3.42110\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018225.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018225    k       Bacteria   99.51678\n",
      "1  YuJ_2015  ERR1018225    k        Archaea    0.46770\n",
      "2  YuJ_2015  ERR1018225    k        Viruses    0.01552\n",
      "3  YuJ_2015  ERR1018225    p  Bacteroidetes   51.65022\n",
      "4  YuJ_2015  ERR1018225    p     Firmicutes   41.36114\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018238.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018238    k        Bacteria   99.99564\n",
      "1  YuJ_2015  ERR1018238    k       Eukaryota    0.00436\n",
      "2  YuJ_2015  ERR1018238    p      Firmicutes   80.79623\n",
      "3  YuJ_2015  ERR1018238    p  Actinobacteria    8.98329\n",
      "4  YuJ_2015  ERR1018238    p   Bacteroidetes    5.87495\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018245.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018245    k        Bacteria   98.45256\n",
      "1  YuJ_2015  ERR1018245    k         Viruses    1.54744\n",
      "2  YuJ_2015  ERR1018245    p      Firmicutes   74.94521\n",
      "3  YuJ_2015  ERR1018245    p   Bacteroidetes   12.91981\n",
      "4  YuJ_2015  ERR1018245    p  Proteobacteria    8.42215\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018253.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018253    k        Bacteria   99.81067\n",
      "1  YuJ_2015  ERR1018253    k         Viruses    0.18933\n",
      "2  YuJ_2015  ERR1018253    p      Firmicutes   59.40649\n",
      "3  YuJ_2015  ERR1018253    p  Actinobacteria   29.99554\n",
      "4  YuJ_2015  ERR1018253    p   Bacteroidetes    7.24463\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018232.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018232    k        Bacteria   99.99474\n",
      "1  YuJ_2015  ERR1018232    k         Viruses    0.00526\n",
      "2  YuJ_2015  ERR1018232    p      Firmicutes   84.43083\n",
      "3  YuJ_2015  ERR1018232    p   Bacteroidetes    7.83106\n",
      "4  YuJ_2015  ERR1018232    p  Proteobacteria    5.48557\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018226.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018226    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018226    p      Firmicutes   72.09079\n",
      "2  YuJ_2015  ERR1018226    p   Bacteroidetes   20.09427\n",
      "3  YuJ_2015  ERR1018226    p  Actinobacteria    6.83744\n",
      "4  YuJ_2015  ERR1018226    p  Proteobacteria    0.73856\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018254.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018254    k        Bacteria   99.48621\n",
      "1  YuJ_2015  ERR1018254    k         Viruses    0.51033\n",
      "2  YuJ_2015  ERR1018254    k         Archaea    0.00345\n",
      "3  YuJ_2015  ERR1018254    p      Firmicutes   59.12180\n",
      "4  YuJ_2015  ERR1018254    p  Actinobacteria   21.61607\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018246.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018246    k        Bacteria   99.89248\n",
      "1  YuJ_2015  ERR1018246    k         Archaea    0.10752\n",
      "2  YuJ_2015  ERR1018246    p      Firmicutes   70.96634\n",
      "3  YuJ_2015  ERR1018246    p   Bacteroidetes   21.54021\n",
      "4  YuJ_2015  ERR1018246    p  Actinobacteria    4.03735\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018239.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018239    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018239    p      Firmicutes   59.79077\n",
      "2  YuJ_2015  ERR1018239    p  Actinobacteria   18.44659\n",
      "3  YuJ_2015  ERR1018239    p   Bacteroidetes   11.58468\n",
      "4  YuJ_2015  ERR1018239    p  Proteobacteria    5.53649\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018233.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018233    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018233    p   Bacteroidetes   67.23039\n",
      "2  YuJ_2015  ERR1018233    p      Firmicutes   23.85322\n",
      "3  YuJ_2015  ERR1018233    p  Actinobacteria    7.34533\n",
      "4  YuJ_2015  ERR1018233    p  Proteobacteria    1.49003\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018255.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018255    k        Bacteria   98.48260\n",
      "1  YuJ_2015  ERR1018255    k         Archaea    1.51740\n",
      "2  YuJ_2015  ERR1018255    p      Firmicutes   43.97810\n",
      "3  YuJ_2015  ERR1018255    p   Bacteroidetes   41.89234\n",
      "4  YuJ_2015  ERR1018255    p  Proteobacteria    6.50889\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018247.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018247    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018247    p  Proteobacteria   59.84718\n",
      "2  YuJ_2015  ERR1018247    p      Firmicutes   20.76366\n",
      "3  YuJ_2015  ERR1018247    p   Bacteroidetes   16.94174\n",
      "4  YuJ_2015  ERR1018247    p  Actinobacteria    2.42746\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018262.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018262    k       Bacteria   99.96704\n",
      "1  YuJ_2015  ERR1018262    k        Viruses    0.03152\n",
      "2  YuJ_2015  ERR1018262    k      Eukaryota    0.00144\n",
      "3  YuJ_2015  ERR1018262    p     Firmicutes   76.32604\n",
      "4  YuJ_2015  ERR1018262    p  Bacteroidetes   16.18404\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018240.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018240    k       Bacteria   99.31314\n",
      "1  YuJ_2015  ERR1018240    k        Archaea    0.48871\n",
      "2  YuJ_2015  ERR1018240    k        Viruses    0.19815\n",
      "3  YuJ_2015  ERR1018240    p     Firmicutes   53.29466\n",
      "4  YuJ_2015  ERR1018240    p  Bacteroidetes   40.89560\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018256.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018256    k        Bacteria   99.60008\n",
      "1  YuJ_2015  ERR1018256    k         Archaea    0.39992\n",
      "2  YuJ_2015  ERR1018256    p      Firmicutes   59.47145\n",
      "3  YuJ_2015  ERR1018256    p   Bacteroidetes   28.22566\n",
      "4  YuJ_2015  ERR1018256    p  Actinobacteria   10.33144\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018269.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018269    k        Bacteria   99.99298\n",
      "1  YuJ_2015  ERR1018269    k       Eukaryota    0.00702\n",
      "2  YuJ_2015  ERR1018269    p   Bacteroidetes   54.19758\n",
      "3  YuJ_2015  ERR1018269    p      Firmicutes   39.69148\n",
      "4  YuJ_2015  ERR1018269    p  Actinobacteria    3.41153\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018276.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018276    k        Bacteria   99.98887\n",
      "1  YuJ_2015  ERR1018276    k         Viruses    0.01113\n",
      "2  YuJ_2015  ERR1018276    p      Firmicutes   55.99512\n",
      "3  YuJ_2015  ERR1018276    p   Bacteroidetes   42.79817\n",
      "4  YuJ_2015  ERR1018276    p  Actinobacteria    0.78537\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018263.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018263    k       Bacteria   99.99531\n",
      "1  YuJ_2015  ERR1018263    k        Viruses    0.00360\n",
      "2  YuJ_2015  ERR1018263    k      Eukaryota    0.00109\n",
      "3  YuJ_2015  ERR1018263    p     Firmicutes   72.41173\n",
      "4  YuJ_2015  ERR1018263    p  Bacteroidetes   18.27172\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018257.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018257    k        Bacteria   99.99570\n",
      "1  YuJ_2015  ERR1018257    k         Viruses    0.00430\n",
      "2  YuJ_2015  ERR1018257    p      Firmicutes   75.90169\n",
      "3  YuJ_2015  ERR1018257    p   Bacteroidetes    9.69240\n",
      "4  YuJ_2015  ERR1018257    p  Proteobacteria    7.96541\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018283.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018283    k        Bacteria   99.99800\n",
      "1  YuJ_2015  ERR1018283    k         Viruses    0.00200\n",
      "2  YuJ_2015  ERR1018283    p      Firmicutes   79.78346\n",
      "3  YuJ_2015  ERR1018283    p  Actinobacteria   12.89007\n",
      "4  YuJ_2015  ERR1018283    p   Bacteroidetes    6.10634\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018277.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018277    k        Bacteria   99.93426\n",
      "1  YuJ_2015  ERR1018277    k         Viruses    0.06574\n",
      "2  YuJ_2015  ERR1018277    p      Firmicutes   80.24674\n",
      "3  YuJ_2015  ERR1018277    p   Bacteroidetes   12.69048\n",
      "4  YuJ_2015  ERR1018277    p  Proteobacteria    3.76905\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018270.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018270    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018270    p   Bacteroidetes   50.90307\n",
      "2  YuJ_2015  ERR1018270    p      Firmicutes   47.08501\n",
      "3  YuJ_2015  ERR1018270    p  Proteobacteria    1.43579\n",
      "4  YuJ_2015  ERR1018270    p  Actinobacteria    0.57267\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018264.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018264    k        Bacteria   99.99765\n",
      "1  YuJ_2015  ERR1018264    k         Archaea    0.00235\n",
      "2  YuJ_2015  ERR1018264    p  Proteobacteria   47.00452\n",
      "3  YuJ_2015  ERR1018264    p      Firmicutes   35.28517\n",
      "4  YuJ_2015  ERR1018264    p   Bacteroidetes   11.57832\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018284.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018284    k        Bacteria   99.99478\n",
      "1  YuJ_2015  ERR1018284    k         Viruses    0.00522\n",
      "2  YuJ_2015  ERR1018284    p      Firmicutes   67.45531\n",
      "3  YuJ_2015  ERR1018284    p  Proteobacteria   16.21399\n",
      "4  YuJ_2015  ERR1018284    p   Bacteroidetes   12.13684\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018258.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018258    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018258    p   Bacteroidetes   45.27806\n",
      "2  YuJ_2015  ERR1018258    p      Firmicutes   38.11480\n",
      "3  YuJ_2015  ERR1018258    p  Actinobacteria   13.90746\n",
      "4  YuJ_2015  ERR1018258    p  Proteobacteria    2.57167\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018278.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018278    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018278    p      Firmicutes   69.48978\n",
      "2  YuJ_2015  ERR1018278    p   Bacteroidetes   20.56687\n",
      "3  YuJ_2015  ERR1018278    p  Proteobacteria    9.72103\n",
      "4  YuJ_2015  ERR1018278    p  Actinobacteria    0.21376\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018271.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018271    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018271    p  Proteobacteria   76.05290\n",
      "2  YuJ_2015  ERR1018271    p      Firmicutes   18.23877\n",
      "3  YuJ_2015  ERR1018271    p   Bacteroidetes    5.16255\n",
      "4  YuJ_2015  ERR1018271    p  Actinobacteria    0.49105\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018285.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018285    k       Bacteria   94.56239\n",
      "1  YuJ_2015  ERR1018285    k        Archaea    5.43761\n",
      "2  YuJ_2015  ERR1018285    p  Bacteroidetes   47.62779\n",
      "3  YuJ_2015  ERR1018285    p     Firmicutes   44.49872\n",
      "4  YuJ_2015  ERR1018285    p  Euryarchaeota    5.43761\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018265.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018265    k        Bacteria   99.99566\n",
      "1  YuJ_2015  ERR1018265    k         Archaea    0.00434\n",
      "2  YuJ_2015  ERR1018265    p   Bacteroidetes   66.06008\n",
      "3  YuJ_2015  ERR1018265    p      Firmicutes   27.24878\n",
      "4  YuJ_2015  ERR1018265    p  Proteobacteria    5.11597\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018259.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018259    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018259    p      Firmicutes   67.21420\n",
      "2  YuJ_2015  ERR1018259    p   Bacteroidetes   20.12758\n",
      "3  YuJ_2015  ERR1018259    p  Proteobacteria    9.53618\n",
      "4  YuJ_2015  ERR1018259    p    Fusobacteria    1.67441\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018279.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018279    k        Bacteria   99.99807\n",
      "1  YuJ_2015  ERR1018279    k         Archaea    0.00193\n",
      "2  YuJ_2015  ERR1018279    p   Bacteroidetes   58.90912\n",
      "3  YuJ_2015  ERR1018279    p  Proteobacteria   29.70232\n",
      "4  YuJ_2015  ERR1018279    p      Firmicutes   11.31571\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018272.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018272    k       Bacteria   97.93935\n",
      "1  YuJ_2015  ERR1018272    k        Archaea    2.03141\n",
      "2  YuJ_2015  ERR1018272    k        Viruses    0.02924\n",
      "3  YuJ_2015  ERR1018272    p  Bacteroidetes   57.63987\n",
      "4  YuJ_2015  ERR1018272    p     Firmicutes   35.64825\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018286.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018286    k       Bacteria   99.99282\n",
      "1  YuJ_2015  ERR1018286    k        Archaea    0.00463\n",
      "2  YuJ_2015  ERR1018286    k        Viruses    0.00255\n",
      "3  YuJ_2015  ERR1018286    p     Firmicutes   69.49834\n",
      "4  YuJ_2015  ERR1018286    p  Bacteroidetes   19.06420\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018260.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018260    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018260    p   Bacteroidetes   50.73766\n",
      "2  YuJ_2015  ERR1018260    p      Firmicutes   45.86602\n",
      "3  YuJ_2015  ERR1018260    p    Fusobacteria    1.78605\n",
      "4  YuJ_2015  ERR1018260    p  Proteobacteria    0.87265\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018266.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018266    k        Bacteria   99.99653\n",
      "1  YuJ_2015  ERR1018266    k         Viruses    0.00347\n",
      "2  YuJ_2015  ERR1018266    p      Firmicutes   50.94980\n",
      "3  YuJ_2015  ERR1018266    p   Bacteroidetes   43.37299\n",
      "4  YuJ_2015  ERR1018266    p  Actinobacteria    2.91453\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018280.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018280    k       Bacteria   99.96998\n",
      "1  YuJ_2015  ERR1018280    k      Eukaryota    0.02582\n",
      "2  YuJ_2015  ERR1018280    k        Viruses    0.00419\n",
      "3  YuJ_2015  ERR1018280    p     Firmicutes   60.51364\n",
      "4  YuJ_2015  ERR1018280    p  Bacteroidetes   31.21834\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018273.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018273    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018273    p      Firmicutes   60.97625\n",
      "2  YuJ_2015  ERR1018273    p   Bacteroidetes   34.68096\n",
      "3  YuJ_2015  ERR1018273    p  Proteobacteria    2.31689\n",
      "4  YuJ_2015  ERR1018273    p  Actinobacteria    1.00218\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018287.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018287    k        Bacteria   99.68248\n",
      "1  YuJ_2015  ERR1018287    k         Viruses    0.31207\n",
      "2  YuJ_2015  ERR1018287    k       Eukaryota    0.00545\n",
      "3  YuJ_2015  ERR1018287    p      Firmicutes   74.84426\n",
      "4  YuJ_2015  ERR1018287    p  Actinobacteria   15.96247\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018261.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018261    k       Bacteria   99.97463\n",
      "1  YuJ_2015  ERR1018261    k        Viruses    0.01662\n",
      "2  YuJ_2015  ERR1018261    k      Eukaryota    0.00874\n",
      "3  YuJ_2015  ERR1018261    p     Firmicutes   73.95128\n",
      "4  YuJ_2015  ERR1018261    p  Bacteroidetes   14.69588\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018267.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018267    k         Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018267    p       Firmicutes   48.10434\n",
      "2  YuJ_2015  ERR1018267    p    Bacteroidetes   39.17260\n",
      "3  YuJ_2015  ERR1018267    p   Actinobacteria    5.76138\n",
      "4  YuJ_2015  ERR1018267    p  Verrucomicrobia    4.82200\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018288.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018288    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018288    p      Firmicutes   55.34907\n",
      "2  YuJ_2015  ERR1018288    p   Bacteroidetes   38.30356\n",
      "3  YuJ_2015  ERR1018288    p  Actinobacteria    3.69943\n",
      "4  YuJ_2015  ERR1018288    p  Proteobacteria    2.62586\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018281.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018281    k         Bacteria   99.97505\n",
      "1  YuJ_2015  ERR1018281    k          Viruses    0.02495\n",
      "2  YuJ_2015  ERR1018281    p       Firmicutes   40.40874\n",
      "3  YuJ_2015  ERR1018281    p    Bacteroidetes   37.27754\n",
      "4  YuJ_2015  ERR1018281    p  Verrucomicrobia   11.29577\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018274.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018274    k        Bacteria   99.99168\n",
      "1  YuJ_2015  ERR1018274    k         Viruses    0.00832\n",
      "2  YuJ_2015  ERR1018274    p      Firmicutes   58.23845\n",
      "3  YuJ_2015  ERR1018274    p   Bacteroidetes   38.44486\n",
      "4  YuJ_2015  ERR1018274    p  Proteobacteria    2.80976\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018268.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018268    k       Bacteria   97.36052\n",
      "1  YuJ_2015  ERR1018268    k        Archaea    2.63608\n",
      "2  YuJ_2015  ERR1018268    k        Viruses    0.00340\n",
      "3  YuJ_2015  ERR1018268    p     Firmicutes   58.25267\n",
      "4  YuJ_2015  ERR1018268    p  Bacteroidetes   37.40072\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018290.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018290    k        Bacteria   99.99326\n",
      "1  YuJ_2015  ERR1018290    k         Viruses    0.00674\n",
      "2  YuJ_2015  ERR1018290    p      Firmicutes   56.19016\n",
      "3  YuJ_2015  ERR1018290    p   Bacteroidetes   38.25649\n",
      "4  YuJ_2015  ERR1018290    p  Proteobacteria    4.45019\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018289.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018289    k       Bacteria   99.93551\n",
      "1  YuJ_2015  ERR1018289    k        Archaea    0.04849\n",
      "2  YuJ_2015  ERR1018289    k      Eukaryota    0.01600\n",
      "3  YuJ_2015  ERR1018289    p  Bacteroidetes   55.09380\n",
      "4  YuJ_2015  ERR1018289    p     Firmicutes   23.85892\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018282.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018282    k         Viruses   63.39895\n",
      "1  YuJ_2015  ERR1018282    k        Bacteria   36.58787\n",
      "2  YuJ_2015  ERR1018282    k         Archaea    0.01318\n",
      "3  YuJ_2015  ERR1018282    p  Viruses_noname   63.39895\n",
      "4  YuJ_2015  ERR1018282    p      Firmicutes   23.48037\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018297.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018297    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018297    p   Bacteroidetes   55.52386\n",
      "2  YuJ_2015  ERR1018297    p      Firmicutes   41.18932\n",
      "3  YuJ_2015  ERR1018297    p  Proteobacteria    1.86318\n",
      "4  YuJ_2015  ERR1018297    p  Actinobacteria    1.21110\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018291.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018291    k       Bacteria   90.75721\n",
      "1  YuJ_2015  ERR1018291    k        Archaea    7.55185\n",
      "2  YuJ_2015  ERR1018291    k        Viruses    1.69094\n",
      "3  YuJ_2015  ERR1018291    p  Bacteroidetes   54.59566\n",
      "4  YuJ_2015  ERR1018291    p     Firmicutes   23.56402\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018275.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018275    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018275    p      Firmicutes   71.77038\n",
      "2  YuJ_2015  ERR1018275    p   Bacteroidetes   23.61004\n",
      "3  YuJ_2015  ERR1018275    p  Actinobacteria    3.76552\n",
      "4  YuJ_2015  ERR1018275    p  Proteobacteria    0.69098\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018304.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018304    k        Bacteria   99.99705\n",
      "1  YuJ_2015  ERR1018304    k         Viruses    0.00295\n",
      "2  YuJ_2015  ERR1018304    p      Firmicutes   77.19300\n",
      "3  YuJ_2015  ERR1018304    p   Bacteroidetes    9.63445\n",
      "4  YuJ_2015  ERR1018304    p  Proteobacteria    9.23037\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018292.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018292    k        Bacteria   99.99069\n",
      "1  YuJ_2015  ERR1018292    k         Archaea    0.00931\n",
      "2  YuJ_2015  ERR1018292    p   Bacteroidetes   54.43781\n",
      "3  YuJ_2015  ERR1018292    p      Firmicutes   42.78342\n",
      "4  YuJ_2015  ERR1018292    p  Proteobacteria    1.34865\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018311.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018311    k        Bacteria   99.91723\n",
      "1  YuJ_2015  ERR1018311    k         Viruses    0.08277\n",
      "2  YuJ_2015  ERR1018311    p      Firmicutes   79.82909\n",
      "3  YuJ_2015  ERR1018311    p   Bacteroidetes   16.11458\n",
      "4  YuJ_2015  ERR1018311    p  Actinobacteria    2.83245\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018298.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018298    k         Bacteria   98.51722\n",
      "1  YuJ_2015  ERR1018298    k          Archaea    1.48278\n",
      "2  YuJ_2015  ERR1018298    p       Firmicutes   60.24118\n",
      "3  YuJ_2015  ERR1018298    p  Verrucomicrobia   19.34580\n",
      "4  YuJ_2015  ERR1018298    p    Bacteroidetes    8.18296\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018293.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018293    k         Bacteria   99.98504\n",
      "1  YuJ_2015  ERR1018293    k          Viruses    0.01337\n",
      "2  YuJ_2015  ERR1018293    k          Archaea    0.00160\n",
      "3  YuJ_2015  ERR1018293    p  Verrucomicrobia   40.73895\n",
      "4  YuJ_2015  ERR1018293    p       Firmicutes   27.11477\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018305.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018305    k       Bacteria   93.94180\n",
      "1  YuJ_2015  ERR1018305    k        Viruses    6.04853\n",
      "2  YuJ_2015  ERR1018305    k        Archaea    0.00967\n",
      "3  YuJ_2015  ERR1018305    p     Firmicutes   46.29105\n",
      "4  YuJ_2015  ERR1018305    p  Bacteroidetes   36.17942\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018299.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018299    k        Bacteria   99.98831\n",
      "1  YuJ_2015  ERR1018299    k         Viruses    0.01169\n",
      "2  YuJ_2015  ERR1018299    p   Bacteroidetes   73.57921\n",
      "3  YuJ_2015  ERR1018299    p      Firmicutes   19.65671\n",
      "4  YuJ_2015  ERR1018299    p  Proteobacteria    3.37514\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018312.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018312    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018312    p  Proteobacteria   47.76052\n",
      "2  YuJ_2015  ERR1018312    p      Firmicutes   27.32900\n",
      "3  YuJ_2015  ERR1018312    p  Actinobacteria   18.93006\n",
      "4  YuJ_2015  ERR1018312    p   Bacteroidetes    5.05994\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018294.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018294    k        Bacteria   99.96135\n",
      "1  YuJ_2015  ERR1018294    k         Viruses    0.03865\n",
      "2  YuJ_2015  ERR1018294    p      Firmicutes   50.76347\n",
      "3  YuJ_2015  ERR1018294    p   Bacteroidetes   38.49588\n",
      "4  YuJ_2015  ERR1018294    p  Actinobacteria    6.13732\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018300.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018300    k        Bacteria   99.98674\n",
      "1  YuJ_2015  ERR1018300    k         Archaea    0.01326\n",
      "2  YuJ_2015  ERR1018300    p   Bacteroidetes   63.95122\n",
      "3  YuJ_2015  ERR1018300    p      Firmicutes   28.33972\n",
      "4  YuJ_2015  ERR1018300    p  Proteobacteria    3.99934\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018306.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018306    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018306    p   Bacteroidetes   65.89279\n",
      "2  YuJ_2015  ERR1018306    p      Firmicutes   27.56788\n",
      "3  YuJ_2015  ERR1018306    p  Actinobacteria    5.28111\n",
      "4  YuJ_2015  ERR1018306    p  Proteobacteria    1.10323\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018295.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018295    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018295    p   Bacteroidetes   66.50391\n",
      "2  YuJ_2015  ERR1018295    p      Firmicutes   31.41088\n",
      "3  YuJ_2015  ERR1018295    p  Proteobacteria    1.53793\n",
      "4  YuJ_2015  ERR1018295    p  Actinobacteria    0.54200\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018301.profile.txt\n",
      "    Dataset   Sample_ID Rank           Name  Abundance\n",
      "0  YuJ_2015  ERR1018301    k       Bacteria   99.98833\n",
      "1  YuJ_2015  ERR1018301    k        Viruses    0.01167\n",
      "2  YuJ_2015  ERR1018301    p  Bacteroidetes   80.35202\n",
      "3  YuJ_2015  ERR1018301    p     Firmicutes   14.03076\n",
      "4  YuJ_2015  ERR1018301    p  Synergistetes    2.48441\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018296.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018296    k        Bacteria   99.93662\n",
      "1  YuJ_2015  ERR1018296    k         Archaea    0.06338\n",
      "2  YuJ_2015  ERR1018296    p   Bacteroidetes   50.82140\n",
      "3  YuJ_2015  ERR1018296    p      Firmicutes   40.50278\n",
      "4  YuJ_2015  ERR1018296    p  Proteobacteria    5.48009\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018307.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018307    k        Bacteria   99.68749\n",
      "1  YuJ_2015  ERR1018307    k         Viruses    0.20844\n",
      "2  YuJ_2015  ERR1018307    k         Archaea    0.10407\n",
      "3  YuJ_2015  ERR1018307    p  Proteobacteria   52.54428\n",
      "4  YuJ_2015  ERR1018307    p      Firmicutes   38.34148\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018302.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018302    k        Bacteria   99.92272\n",
      "1  YuJ_2015  ERR1018302    k         Viruses    0.07728\n",
      "2  YuJ_2015  ERR1018302    p   Bacteroidetes   48.94252\n",
      "3  YuJ_2015  ERR1018302    p      Firmicutes   29.62431\n",
      "4  YuJ_2015  ERR1018302    p  Proteobacteria   13.47387\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018308.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018308    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018308    p      Firmicutes   87.45732\n",
      "2  YuJ_2015  ERR1018308    p   Bacteroidetes    9.66131\n",
      "3  YuJ_2015  ERR1018308    p  Actinobacteria    1.75166\n",
      "4  YuJ_2015  ERR1018308    p  Proteobacteria    0.96311\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018303.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018303    k        Bacteria   99.99717\n",
      "1  YuJ_2015  ERR1018303    k         Viruses    0.00283\n",
      "2  YuJ_2015  ERR1018303    p      Firmicutes   51.86971\n",
      "3  YuJ_2015  ERR1018303    p   Bacteroidetes   45.00152\n",
      "4  YuJ_2015  ERR1018303    p  Proteobacteria    1.64708\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018309.profile.txt\n",
      "    Dataset   Sample_ID Rank             Name  Abundance\n",
      "0  YuJ_2015  ERR1018309    k         Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018309    p       Firmicutes   78.73796\n",
      "2  YuJ_2015  ERR1018309    p   Actinobacteria   11.61668\n",
      "3  YuJ_2015  ERR1018309    p   Proteobacteria    4.24819\n",
      "4  YuJ_2015  ERR1018309    p  Verrucomicrobia    3.68544\n",
      "//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018310.profile.txt\n",
      "    Dataset   Sample_ID Rank            Name  Abundance\n",
      "0  YuJ_2015  ERR1018310    k        Bacteria  100.00000\n",
      "1  YuJ_2015  ERR1018310    p      Firmicutes   56.73464\n",
      "2  YuJ_2015  ERR1018310    p   Bacteroidetes   36.25653\n",
      "3  YuJ_2015  ERR1018310    p  Actinobacteria    5.14026\n",
      "4  YuJ_2015  ERR1018310    p  Proteobacteria    1.84492\n",
      "Files in dataset: 128\n",
      "YuJ_2015  kész.\n",
      "ZellerG_2014  kész.\n"
     ]
    }
   ],
   "source": [
    "dataset_dfs = {}\n",
    "i = 0;\n",
    "for dataset_name, filelist in files.items():\n",
    "    if dataset_name == \"YuJ_2015\":\n",
    "        df_list = []\n",
    "        fc = 0\n",
    "        for file in filelist:\n",
    "            print(file)\n",
    "            df = get_data_for_sample(file)\n",
    "            print(df.head())\n",
    "            if df is not None:\n",
    "                df_list.append(get_data_for_sample(file))\n",
    "            fc = fc+1\n",
    "        df_dataset = pd.concat(df_list)\n",
    "        output = join_metadata(df_dataset, dataset_name)\n",
    "        print(\"Files in dataset:\", fc)\n",
    "        \n",
    "        path_to_output = os.path.join(path, dataset_name + \".csv\")\n",
    "        with open(path_to_output, \"w\") as f: \n",
    "            f.write(output.to_csv()) \n",
    "    i = i+1\n",
    "    print(dataset_name, \" kész.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import seaborn as sns\n",
    "from skbio.diversity import alpha_diversity\n",
    "from skbio.diversity import beta_diversity\n",
    "from skbio import TreeNode\n",
    "from io import StringIO\n",
    "from skbio.stats.ordination import pcoa\n",
    "\n",
    "import cx_Oracle\n",
    "\n",
    "SERVER = 'dboracle.itk.ppke.hu'\n",
    "PORT = 1521\n",
    "SERVICE = 'rsc.itk.ppke.hu'\n",
    "USER = 'metagenome'\n",
    "PASSWORD = 'LrJPRUS73r84'\n",
    "\n",
    "class Connection():\n",
    "\n",
    "    def __init__(self):\n",
    "        dsn_tns = cx_Oracle.makedsn(SERVER, PORT, service_name=SERVICE)\n",
    "        self.connection = cx_Oracle.connect(USER, PASSWORD, dsn_tns)\n",
    "\n",
    "    def get_connection(self):\n",
    "        return self.connection\n",
    "\n",
    "    def select(self, sql):\n",
    "        return pd.read_sql(sql, con=self.connection)\n",
    "\n",
    "con = Connection()\n",
    "\n",
    "path = '//gfs/data/curated_metagenomes_metaphlan2/'\n",
    "\n",
    "files = {}\n",
    "# r=root, d=directories, f = files\n",
    "for r, d, f in os.walk(path):\n",
    "    for directory in d:\n",
    "        for r1, d1, f1 in os.walk(os.path.join(path,directory)):\n",
    "            for file in f1:\n",
    "                if '.profile.txt' in file:\n",
    "                    if directory in files:\n",
    "                        files[directory].append(os.path.join(r1, file))\n",
    "                    else:\n",
    "                        files[directory] = [(os.path.join(r1, file))]\n",
    "\n",
    "def get_data_for_sample(path_to_file):\n",
    "    dataset = path_to_file.split('/')[-2]\n",
    "    sample_id = path_to_file.split('/')[-1].split('.')[0]\n",
    "\n",
    "    df = pd.read_csv(path_to_file, delimiter = '\\t')\n",
    "\n",
    "    profile_df_2 = df\n",
    "    profile_df_2['Rank'] = None\n",
    "    profile_df_2['K'] = None\n",
    "    profile_df_2['P'] = None\n",
    "    profile_df_2['C'] = None\n",
    "    profile_df_2['O'] = None\n",
    "    profile_df_2['F'] = None\n",
    "    profile_df_2['G'] = None\n",
    "    profile_df_2['S'] = None\n",
    "    profile_df_2['T'] = None\n",
    "    profile_df_2['Abundance'] = profile_df_2[\"Metaphlan2_Analysis\"]\n",
    "\n",
    "    for p in profile_df_2[\"#SampleID\"]:\n",
    "        row_level = None\n",
    "        tags = p.split('|')\n",
    "        for tag in tags:\n",
    "            tmp = tag.split('__')\n",
    "            try:\n",
    "                level = tmp[0]\n",
    "                name = tmp[1]\n",
    "                profile_df_2\n",
    "                row_level = level\n",
    "\n",
    "                profile_df_2.loc[profile_df_2['#SampleID'] == p, level] = name\n",
    "            except:\n",
    "                print(\"Can not split:\", tmp)\n",
    "        profile_df_2.loc[profile_df_2['#SampleID'] == p, 'Rank'] = row_level\n",
    "\n",
    "    def get_name_for_row(row):\n",
    "        try:\n",
    "            return row[row.Rank]\n",
    "        except:\n",
    "            return '???'\n",
    "\n",
    "    profile_df_2[\"Name\"] = profile_df_2.apply(get_name_for_row, axis=1)\n",
    "\n",
    "    selected = profile_df_2[[\"Rank\", \"Name\", \"Abundance\"]]\n",
    "\n",
    "    selected['Sample_ID'] = sample_id\n",
    "    selected['Dataset'] = dataset\n",
    "\n",
    "\n",
    "    return selected[[\"Dataset\", \"Sample_ID\", \"Rank\", \"Name\", \"Abundance\"]]\n",
    "\n",
    "def join_metadata(df, dataset_name):\n",
    "    SQL_METADATA = \"\"\"\n",
    "        select\n",
    "            DATASET_NAME,\n",
    "            STUDY_CONDITION,\n",
    "            BODY_SITE,\n",
    "            SUBJECTID,\n",
    "            NCBI_ACCESSION,\n",
    "            COUNTRY,\n",
    "            DISEASE,\n",
    "            GENDER,\n",
    "            AGE,\n",
    "            AGE_CATEGORY\n",
    "        from METADATA\n",
    "        where dataset_name = '{dataset_name}'\n",
    "        \"\"\".replace(\"{dataset_name}\", dataset_name)\n",
    "    df_metadata = con.select(SQL_METADATA)\n",
    "\n",
    "    df[\"DATASET_NAME\"] = df[\"Dataset\"]\n",
    "    merged = df.merge(df_metadata, on='DATASET_NAME')\n",
    "\n",
    "    col_to_idx = {}\n",
    "    for i in range(len(merged.columns)):\n",
    "        col_to_idx[merged.columns[i]] = i\n",
    "\n",
    "    joined = merged.loc[[row[1][col_to_idx[\"Sample_ID\"]] in row[1][col_to_idx[\"NCBI_ACCESSION\"]]\n",
    "                        for row in merged.iterrows()]]\n",
    "\n",
    "    output = joined[[\"Sample_ID\", \"Dataset\", \"Rank\", \"Name\", \"Abundance\",\n",
    "                     \"STUDY_CONDITION\", \"BODY_SITE\", \"SUBJECTID\",\n",
    "                     \"COUNTRY\", \"DISEASE\", \"AGE\", \"AGE_CATEGORY\"]]\n",
    "    output = output.reset_index()\n",
    "    output = output.drop(columns=\"index\")\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018191.profile.txt'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files[\"YuJ_2015\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gfs/progs/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:90: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/gfs/progs/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:91: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "df = get_data_for_sample('//gfs/data/curated_metagenomes_metaphlan2/YuJ_2015/ERR1018191.profile.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Name</th>\n",
       "      <th>Abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>99.98996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>k</td>\n",
       "      <td>Viruses</td>\n",
       "      <td>0.01004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>p</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>58.59706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>p</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>18.78107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>p</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>18.20503</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Dataset   Sample_ID Rank            Name  Abundance\n",
       "0  YuJ_2015  ERR1018191    k        Bacteria   99.98996\n",
       "1  YuJ_2015  ERR1018191    k         Viruses    0.01004\n",
       "2  YuJ_2015  ERR1018191    p      Firmicutes   58.59706\n",
       "3  YuJ_2015  ERR1018191    p  Actinobacteria   18.78107\n",
       "4  YuJ_2015  ERR1018191    p  Proteobacteria   18.20503"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Name</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Sample_ID, Dataset, Rank, Name, Abundance, STUDY_CONDITION, BODY_SITE, SUBJECTID, COUNTRY, DISEASE, AGE, AGE_CATEGORY]\n",
       "Index: []"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined = join_metadata(df, \"YuJ_2015\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL_METADATA = \"\"\"\n",
    "    select\n",
    "        DATASET_NAME,\n",
    "        STUDY_CONDITION,\n",
    "        BODY_SITE,\n",
    "        SUBJECTID,\n",
    "        NCBI_ACCESSION,\n",
    "        COUNTRY,\n",
    "        DISEASE,\n",
    "        GENDER,\n",
    "        AGE,\n",
    "        AGE_CATEGORY\n",
    "    from METADATA\n",
    "    where dataset_name = '{dataset_name}'\n",
    "    \"\"\".replace(\"{dataset_name}\", \"YuJ_2015\")\n",
    "df_metadata = con.select(SQL_METADATA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATASET_NAME</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>NCBI_ACCESSION</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>CRC</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017595-169</td>\n",
       "      <td>ERS848745</td>\n",
       "      <td>CHN</td>\n",
       "      <td>CRC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>CRC</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017594-166</td>\n",
       "      <td>ERS848744</td>\n",
       "      <td>CHN</td>\n",
       "      <td>CRC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>CRC</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017593-158</td>\n",
       "      <td>ERS848743</td>\n",
       "      <td>CHN</td>\n",
       "      <td>CRC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017592-142</td>\n",
       "      <td>ERS848742</td>\n",
       "      <td>CHN</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017591-140</td>\n",
       "      <td>ERS848741</td>\n",
       "      <td>CHN</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATASET_NAME STUDY_CONDITION BODY_SITE         SUBJECTID NCBI_ACCESSION  \\\n",
       "0     YuJ_2015             CRC     stool  SZAXPI017595-169      ERS848745   \n",
       "1     YuJ_2015             CRC     stool  SZAXPI017594-166      ERS848744   \n",
       "2     YuJ_2015             CRC     stool  SZAXPI017593-158      ERS848743   \n",
       "3     YuJ_2015         control     stool  SZAXPI017592-142      ERS848742   \n",
       "4     YuJ_2015         control     stool  SZAXPI017591-140      ERS848741   \n",
       "\n",
       "  COUNTRY DISEASE GENDER   AGE AGE_CATEGORY  \n",
       "0     CHN     CRC   None  None       senior  \n",
       "1     CHN     CRC   None  None       senior  \n",
       "2     CHN     CRC   None  None       senior  \n",
       "3     CHN    none   None  None       senior  \n",
       "4     CHN    none   None  None       senior  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"DATASET_NAME\"] = df[\"Dataset\"]\n",
    "merged = df.merge(df_metadata, on='DATASET_NAME')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Name</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>DATASET_NAME</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>NCBI_ACCESSION</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>99.98996</td>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>CRC</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017595-169</td>\n",
       "      <td>ERS848745</td>\n",
       "      <td>CHN</td>\n",
       "      <td>CRC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>99.98996</td>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>CRC</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017594-166</td>\n",
       "      <td>ERS848744</td>\n",
       "      <td>CHN</td>\n",
       "      <td>CRC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>99.98996</td>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>CRC</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017593-158</td>\n",
       "      <td>ERS848743</td>\n",
       "      <td>CHN</td>\n",
       "      <td>CRC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>99.98996</td>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017592-142</td>\n",
       "      <td>ERS848742</td>\n",
       "      <td>CHN</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>ERR1018191</td>\n",
       "      <td>k</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>99.98996</td>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>control</td>\n",
       "      <td>stool</td>\n",
       "      <td>SZAXPI017591-140</td>\n",
       "      <td>ERS848741</td>\n",
       "      <td>CHN</td>\n",
       "      <td>none</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>senior</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Dataset   Sample_ID Rank      Name  Abundance DATASET_NAME  \\\n",
       "0  YuJ_2015  ERR1018191    k  Bacteria   99.98996     YuJ_2015   \n",
       "1  YuJ_2015  ERR1018191    k  Bacteria   99.98996     YuJ_2015   \n",
       "2  YuJ_2015  ERR1018191    k  Bacteria   99.98996     YuJ_2015   \n",
       "3  YuJ_2015  ERR1018191    k  Bacteria   99.98996     YuJ_2015   \n",
       "4  YuJ_2015  ERR1018191    k  Bacteria   99.98996     YuJ_2015   \n",
       "\n",
       "  STUDY_CONDITION BODY_SITE         SUBJECTID NCBI_ACCESSION COUNTRY DISEASE  \\\n",
       "0             CRC     stool  SZAXPI017595-169      ERS848745     CHN     CRC   \n",
       "1             CRC     stool  SZAXPI017594-166      ERS848744     CHN     CRC   \n",
       "2             CRC     stool  SZAXPI017593-158      ERS848743     CHN     CRC   \n",
       "3         control     stool  SZAXPI017592-142      ERS848742     CHN    none   \n",
       "4         control     stool  SZAXPI017591-140      ERS848741     CHN    none   \n",
       "\n",
       "  GENDER   AGE AGE_CATEGORY  \n",
       "0   None  None       senior  \n",
       "1   None  None       senior  \n",
       "2   None  None       senior  \n",
       "3   None  None       senior  \n",
       "4   None  None       senior  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_to_idx = {}\n",
    "for i in range(len(merged.columns)):\n",
    "    col_to_idx[merged.columns[i]] = i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Dataset': 0,\n",
       " 'Sample_ID': 1,\n",
       " 'Rank': 2,\n",
       " 'Name': 3,\n",
       " 'Abundance': 4,\n",
       " 'DATASET_NAME': 5,\n",
       " 'STUDY_CONDITION': 6,\n",
       " 'BODY_SITE': 7,\n",
       " 'SUBJECTID': 8,\n",
       " 'NCBI_ACCESSION': 9,\n",
       " 'COUNTRY': 10,\n",
       " 'DISEASE': 11,\n",
       " 'GENDER': 12,\n",
       " 'AGE': 13,\n",
       " 'AGE_CATEGORY': 14}"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_to_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "row = [row for row in merged.iterrows()][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "row = row[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Dataset                    YuJ_2015\n",
       "Sample_ID                ERR1018191\n",
       "Rank                              k\n",
       "Name                       Bacteria\n",
       "Abundance                     99.99\n",
       "DATASET_NAME               YuJ_2015\n",
       "STUDY_CONDITION                 CRC\n",
       "BODY_SITE                     stool\n",
       "SUBJECTID          SZAXPI017595-169\n",
       "NCBI_ACCESSION            ERS848745\n",
       "COUNTRY                         CHN\n",
       "DISEASE                         CRC\n",
       "GENDER                         None\n",
       "AGE                            None\n",
       "AGE_CATEGORY                 senior\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ERR1018191'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row[col_to_idx[\"Sample_ID\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ERS848745'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row[col_to_idx[\"NCBI_ACCESSION\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "joined = merged.loc[[row[1][col_to_idx[\"Sample_ID\"]] in row[1][col_to_idx[\"NCBI_ACCESSION\"]]\n",
    "                    for row in merged.iterrows()]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Name</th>\n",
       "      <th>Abundance</th>\n",
       "      <th>DATASET_NAME</th>\n",
       "      <th>STUDY_CONDITION</th>\n",
       "      <th>BODY_SITE</th>\n",
       "      <th>SUBJECTID</th>\n",
       "      <th>NCBI_ACCESSION</th>\n",
       "      <th>COUNTRY</th>\n",
       "      <th>DISEASE</th>\n",
       "      <th>GENDER</th>\n",
       "      <th>AGE</th>\n",
       "      <th>AGE_CATEGORY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Dataset, Sample_ID, Rank, Name, Abundance, DATASET_NAME, STUDY_CONDITION, BODY_SITE, SUBJECTID, NCBI_ACCESSION, COUNTRY, DISEASE, GENDER, AGE, AGE_CATEGORY]\n",
       "Index: []"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "dataset_dfs = {}\n",
    "i = 0;\n",
    "for dataset_name, filelist in files.items():\n",
    "    if dataset_name == \"YuJ_2015\":\n",
    "        df_list = []\n",
    "        for file in filelist:\n",
    "            print(file)\n",
    "            df = get_data_for_sample(file)\n",
    "            if df is not None:\n",
    "                df_list.append(get_data_for_sample(file))\n",
    "        df_dataset = pd.concat(df_list)\n",
    "        output = join_metadata(df_dataset, dataset_name)\n",
    "\n",
    "        path_to_output = os.path.join(path, dataset_name + \".csv\")\n",
    "        with open(path_to_output, \"w\") as f:\n",
    "            f.write(output.to_csv())\n",
    "    i = i+1\n",
    "    print(dataset_name, \" kesz.\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
